<!DOCTYPE html>
<html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Chapter 14 Cell cycle Assignment | scRNAseq Analysis in R with Seurat</title>
<meta name="author" content="Monash Genomics and Bioinformatics Platform (MGBP)">
<meta name="description" content="In some datasets, the phase of cell cycle that a cell is in (G1/G2M/S) can account for alot of the observed transcriptomic variation. There may be clustering by phase, or separation in the UMAP by...">
<meta name="generator" content="bookdown 0.41 with bs4_book()">
<meta property="og:title" content="Chapter 14 Cell cycle Assignment | scRNAseq Analysis in R with Seurat">
<meta property="og:type" content="book">
<meta property="og:description" content="In some datasets, the phase of cell cycle that a cell is in (G1/G2M/S) can account for alot of the observed transcriptomic variation. There may be clustering by phase, or separation in the UMAP by...">
<meta name="twitter:card" content="summary">
<meta name="twitter:title" content="Chapter 14 Cell cycle Assignment | scRNAseq Analysis in R with Seurat">
<meta name="twitter:description" content="In some datasets, the phase of cell cycle that a cell is in (G1/G2M/S) can account for alot of the observed transcriptomic variation. There may be clustering by phase, or separation in the UMAP by...">
<!-- JS --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://kit.fontawesome.com/6ecbd6c532.js" crossorigin="anonymous"></script><script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="libs/bootstrap-4.6.0/bootstrap.min.css" rel="stylesheet">
<script src="libs/bootstrap-4.6.0/bootstrap.bundle.min.js"></script><script src="libs/bs3compat-0.8.0/transition.js"></script><script src="libs/bs3compat-0.8.0/tabs.js"></script><script src="libs/bs3compat-0.8.0/bs3compat.js"></script><link href="libs/bs4_book-1.0.0/bs4_book.css" rel="stylesheet">
<script src="libs/bs4_book-1.0.0/bs4_book.js"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- CSS --><style type="text/css">
    
    div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  </style>
<link rel="stylesheet" href="style.css">
</head>
<body data-spy="scroll" data-target="#toc">

<div class="container-fluid">
<div class="row">
  <header class="col-sm-12 col-lg-3 sidebar sidebar-book"><a class="sr-only sr-only-focusable" href="#content">Skip to main content</a>

    <div class="d-flex align-items-start justify-content-between">
      <h1>
        <a href="index.html" title="">scRNAseq Analysis in R with Seurat</a>
      </h1>
      <button class="btn btn-outline-primary d-lg-none ml-2 mt-1" type="button" data-toggle="collapse" data-target="#main-nav" aria-expanded="true" aria-controls="main-nav"><i class="fas fa-bars"></i><span class="sr-only">Show table of contents</span></button>
    </div>

    <div id="main-nav" class="collapse-lg">
      <form role="search">
        <input id="search" class="form-control" type="search" placeholder="Search" aria-label="Search">
</form>

      <nav aria-label="Table of contents"><h2>Table of contents</h2>
        <ul class="book-toc list-unstyled">
<li><a class="" href="index.html"><span class="header-section-number">1</span> Getting started</a></li>
<li><a class="" href="set-up.html"><span class="header-section-number">2</span> Set up</a></li>
<li><a class="" href="schedule.html"><span class="header-section-number">3</span> Schedule</a></li>
<li><a class="" href="load.html"><span class="header-section-number">4</span> Load data</a></li>
<li class="book-part">Single Cell Analysis</li>
<li><a class="" href="qc.html"><span class="header-section-number">5</span> QC Filtering</a></li>
<li><a class="" href="norm.html"><span class="header-section-number">6</span> Normalisation</a></li>
<li><a class="" href="reducedims.html"><span class="header-section-number">7</span> PCAs and UMAPs</a></li>
<li><a class="" href="dimensionality-reduction.html"><span class="header-section-number">8</span> Dimensionality reduction</a></li>
<li><a class="" href="Harmony.html"><span class="header-section-number">9</span> Data set integration with Harmony</a></li>
<li><a class="" href="clustering.html"><span class="header-section-number">10</span> Clustering</a></li>
<li><a class="" href="clustermarkers.html"><span class="header-section-number">11</span> Cluster Markers</a></li>
<li class="book-part">Futher Analysis</li>
<li><a class="" href="singler.html"><span class="header-section-number">12</span> SingleR</a></li>
<li><a class="" href="de2.html"><span class="header-section-number">13</span> Differential Expression</a></li>
<li><a class="active" href="CellCycle.html"><span class="header-section-number">14</span> Cell cycle Assignment</a></li>
<li class="book-part">Other Resources</li>
<li><a class="" href="resources.html"><span class="header-section-number">15</span> Resources</a></li>
<li class="book-part">Seurat Object</li>
<li><a class="" href="structure.html"><span class="header-section-number">16</span> Structure</a></li>
<li><a class="" href="acknowledgements.html"><span class="header-section-number">17</span> Acknowledgements</a></li>
<li><a class="" href="session-info.html"><span class="header-section-number">18</span> Session info</a></li>
</ul>

        <div class="book-extra">
          <p><a id="book-repo" href="https://monashbioinformaticsplatform.github.io/scRNAseq_Workshop_ABACBS_2024/">View book source <i class="fab fa-gitlab"></i></a></p>
        </div>
      </nav>
</div>
  </header><main class="col-sm-12 col-md-9 col-lg-7" id="content"><div id="CellCycle" class="section level1" number="14">
<h1>
<span class="header-section-number">14</span> Cell cycle Assignment<a class="anchor" aria-label="anchor" href="#CellCycle"><i class="fas fa-link"></i></a>
</h1>
<p>In some datasets, the phase of cell cycle that a cell is in (G1/G2M/S) can account for
alot of the observed transcriptomic variation. There may be clustering by phase, or
separation in the UMAP by phase.</p>
<p>Seurat provides a simple method for assigning cell cycle state to each cell. Other methods are available.</p>
<p>More information about assigning cell cycle states to cells is in the <a href="https://satijalab.org/seurat/articles/cell_cycle_vignette.html">cell cycle vignette</a></p>
<div class="sourceCode" id="cb99"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># A list of cell cycle markers, from Tirosh et al, 2015, is loaded with Seurat.  We can</span></span>
<span><span class="co"># segregate this list into markers of G2/M phase and markers of S phase</span></span>
<span><span class="va">s.genes</span>   <span class="op">&lt;-</span> <span class="va">cc.genes</span><span class="op">$</span><span class="va">s.genes</span></span>
<span><span class="va">g2m.genes</span> <span class="op">&lt;-</span> <span class="va">cc.genes</span><span class="op">$</span><span class="va">g2m.genes</span></span>
<span></span>
<span><span class="co"># Use those lists with the cell cycle scoring function in Seurat.</span></span>
<span><span class="va">seurat_object</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/CellCycleScoring.html">CellCycleScoring</a></span><span class="op">(</span><span class="va">seurat_object</span>, s.features <span class="op">=</span> <span class="va">s.genes</span>, g2m.features <span class="op">=</span> <span class="va">g2m.genes</span><span class="op">)</span></span>
<span><span class="co">#&gt; Warning: The following features are not present in the</span></span>
<span><span class="co">#&gt; object: TYMS, MCM2, MCM4, UNG, GINS2, CDCA7, DTL, UHRF1,</span></span>
<span><span class="co">#&gt; MLF1IP, HELLS, RAD51AP1, GMNN, WDR76, CCNE2, ATAD2, RAD51,</span></span>
<span><span class="co">#&gt; RRM2, CDC45, CDC6, EXO1, DSCC1, BLM, CASP8AP2, CLSPN,</span></span>
<span><span class="co">#&gt; POLA1, CHAF1B, BRIP1, E2F8, not searching for symbol</span></span>
<span><span class="co">#&gt; synonyms</span></span>
<span><span class="co">#&gt; Warning: The following features are not present in the</span></span>
<span><span class="co">#&gt; object: CDK1, NUSAP1, UBE2C, BIRC5, TPX2, TOP2A, NDC80,</span></span>
<span><span class="co">#&gt; NUF2, MKI67, FAM64A, CCNB2, CKAP2L, AURKB, BUB1, KIF11,</span></span>
<span><span class="co">#&gt; GTSE1, HJURP, CDCA3, CDC20, TTK, CDC25C, KIF2C, NCAPD2,</span></span>
<span><span class="co">#&gt; DLGAP5, CDCA2, CDCA8, ECT2, KIF23, HMMR, AURKA, PSRC1,</span></span>
<span><span class="co">#&gt; ANLN, CENPE, NEK2, GAS2L3, CENPA, not searching for symbol</span></span>
<span><span class="co">#&gt; synonyms</span></span></code></pre></div>
<p>Which adds S.Score, G2M.Score and Phase calls to the metadata.</p>
<div class="sourceCode" id="cb100"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">seurat_object</span><span class="op">@</span><span class="va">meta.data</span><span class="op">)</span></span>
<span><span class="co">#&gt;                     orig.ident nCount_RNA nFeature_RNA  ind</span></span>
<span><span class="co">#&gt; AGGGCGCTATTTCC-1 SeuratProject       2053          532 1256</span></span>
<span><span class="co">#&gt; GGAGACGATTCGTT-1 SeuratProject        881          392 1256</span></span>
<span><span class="co">#&gt; CACCGTTGTCGTAG-1 SeuratProject       3130         1005 1016</span></span>
<span><span class="co">#&gt; TATCGTACACGCAT-1 SeuratProject       1042          549 1488</span></span>
<span><span class="co">#&gt; TACGAGACCTATTC-1 SeuratProject       2425          777 1244</span></span>
<span><span class="co">#&gt; GTACTACTCATACG-1 SeuratProject       3951         1064 1256</span></span>
<span><span class="co">#&gt;                  stim              cell multiplets</span></span>
<span><span class="co">#&gt; AGGGCGCTATTTCC-1 stim   CD14+ Monocytes    singlet</span></span>
<span><span class="co">#&gt; GGAGACGATTCGTT-1 stim       CD4 T cells    singlet</span></span>
<span><span class="co">#&gt; CACCGTTGTCGTAG-1 ctrl FCGR3A+ Monocytes    singlet</span></span>
<span><span class="co">#&gt; TATCGTACACGCAT-1 stim           B cells    singlet</span></span>
<span><span class="co">#&gt; TACGAGACCTATTC-1 stim       CD4 T cells    singlet</span></span>
<span><span class="co">#&gt; GTACTACTCATACG-1 ctrl FCGR3A+ Monocytes    singlet</span></span>
<span><span class="co">#&gt;                  percent.mt RNA_snn_res.0.25</span></span>
<span><span class="co">#&gt; AGGGCGCTATTTCC-1  1.6336634                1</span></span>
<span><span class="co">#&gt; GGAGACGATTCGTT-1  4.8809524                0</span></span>
<span><span class="co">#&gt; CACCGTTGTCGTAG-1  1.0655473                4</span></span>
<span><span class="co">#&gt; TATCGTACACGCAT-1  3.0662710                3</span></span>
<span><span class="co">#&gt; TACGAGACCTATTC-1  1.0837849                0</span></span>
<span><span class="co">#&gt; GTACTACTCATACG-1  0.7137395                4</span></span>
<span><span class="co">#&gt;                  seurat_clusters pca_clusters</span></span>
<span><span class="co">#&gt; AGGGCGCTATTTCC-1               1            3</span></span>
<span><span class="co">#&gt; GGAGACGATTCGTT-1              12            0</span></span>
<span><span class="co">#&gt; CACCGTTGTCGTAG-1              15            7</span></span>
<span><span class="co">#&gt; TATCGTACACGCAT-1               8            5</span></span>
<span><span class="co">#&gt; TACGAGACCTATTC-1               7            0</span></span>
<span><span class="co">#&gt; GTACTACTCATACG-1              15            7</span></span>
<span><span class="co">#&gt;                  harmony_clusters RNA_snn_res.0.5</span></span>
<span><span class="co">#&gt; AGGGCGCTATTTCC-1                1               3</span></span>
<span><span class="co">#&gt; GGAGACGATTCGTT-1                0               0</span></span>
<span><span class="co">#&gt; CACCGTTGTCGTAG-1                4               9</span></span>
<span><span class="co">#&gt; TATCGTACACGCAT-1                3               6</span></span>
<span><span class="co">#&gt; TACGAGACCTATTC-1                0               0</span></span>
<span><span class="co">#&gt; GTACTACTCATACG-1                4               9</span></span>
<span><span class="co">#&gt;                  RNA_snn_res.0.1 RNA_snn_res.0.2</span></span>
<span><span class="co">#&gt; AGGGCGCTATTTCC-1               3               3</span></span>
<span><span class="co">#&gt; GGAGACGATTCGTT-1               1               0</span></span>
<span><span class="co">#&gt; CACCGTTGTCGTAG-1               2               7</span></span>
<span><span class="co">#&gt; TATCGTACACGCAT-1               4               5</span></span>
<span><span class="co">#&gt; TACGAGACCTATTC-1               1               0</span></span>
<span><span class="co">#&gt; GTACTACTCATACG-1               2               7</span></span>
<span><span class="co">#&gt;                  RNA_snn_res.0.3 RNA_snn_res.0.4</span></span>
<span><span class="co">#&gt; AGGGCGCTATTTCC-1               2               3</span></span>
<span><span class="co">#&gt; GGAGACGATTCGTT-1               0               0</span></span>
<span><span class="co">#&gt; CACCGTTGTCGTAG-1               9               9</span></span>
<span><span class="co">#&gt; TATCGTACACGCAT-1               6               6</span></span>
<span><span class="co">#&gt; TACGAGACCTATTC-1               0               0</span></span>
<span><span class="co">#&gt; GTACTACTCATACG-1               9               9</span></span>
<span><span class="co">#&gt;                  RNA_snn_res.0.6 RNA_snn_res.0.7</span></span>
<span><span class="co">#&gt; AGGGCGCTATTTCC-1               3               3</span></span>
<span><span class="co">#&gt; GGAGACGATTCGTT-1               0               0</span></span>
<span><span class="co">#&gt; CACCGTTGTCGTAG-1               9               9</span></span>
<span><span class="co">#&gt; TATCGTACACGCAT-1               6               6</span></span>
<span><span class="co">#&gt; TACGAGACCTATTC-1               0               0</span></span>
<span><span class="co">#&gt; GTACTACTCATACG-1               9               9</span></span>
<span><span class="co">#&gt;                  RNA_snn_res.0.8 RNA_snn_res.0.9</span></span>
<span><span class="co">#&gt; AGGGCGCTATTTCC-1               3               3</span></span>
<span><span class="co">#&gt; GGAGACGATTCGTT-1               0               9</span></span>
<span><span class="co">#&gt; CACCGTTGTCGTAG-1               9              10</span></span>
<span><span class="co">#&gt; TATCGTACACGCAT-1               6               6</span></span>
<span><span class="co">#&gt; TACGAGACCTATTC-1               0               0</span></span>
<span><span class="co">#&gt; GTACTACTCATACG-1               9              10</span></span>
<span><span class="co">#&gt;                  RNA_snn_res.1 RNA_snn_res.1.1</span></span>
<span><span class="co">#&gt; AGGGCGCTATTTCC-1             3               3</span></span>
<span><span class="co">#&gt; GGAGACGATTCGTT-1            10              10</span></span>
<span><span class="co">#&gt; CACCGTTGTCGTAG-1            11              11</span></span>
<span><span class="co">#&gt; TATCGTACACGCAT-1             6               6</span></span>
<span><span class="co">#&gt; TACGAGACCTATTC-1             0               0</span></span>
<span><span class="co">#&gt; GTACTACTCATACG-1            11              11</span></span>
<span><span class="co">#&gt;                  RNA_snn_res.1.2 RNA_snn_res.1.3</span></span>
<span><span class="co">#&gt; AGGGCGCTATTTCC-1               2               2</span></span>
<span><span class="co">#&gt; GGAGACGATTCGTT-1              11              12</span></span>
<span><span class="co">#&gt; CACCGTTGTCGTAG-1              12              13</span></span>
<span><span class="co">#&gt; TATCGTACACGCAT-1               6               7</span></span>
<span><span class="co">#&gt; TACGAGACCTATTC-1               0               0</span></span>
<span><span class="co">#&gt; GTACTACTCATACG-1              12              13</span></span>
<span><span class="co">#&gt;                  RNA_snn_res.1.4 RNA_snn_res.1.5</span></span>
<span><span class="co">#&gt; AGGGCGCTATTTCC-1               2               2</span></span>
<span><span class="co">#&gt; GGAGACGATTCGTT-1              13              13</span></span>
<span><span class="co">#&gt; CACCGTTGTCGTAG-1              14              14</span></span>
<span><span class="co">#&gt; TATCGTACACGCAT-1               8               8</span></span>
<span><span class="co">#&gt; TACGAGACCTATTC-1               0               0</span></span>
<span><span class="co">#&gt; GTACTACTCATACG-1              14              14</span></span>
<span><span class="co">#&gt;                  RNA_snn_res.1.6 RNA_snn_res.1.7</span></span>
<span><span class="co">#&gt; AGGGCGCTATTTCC-1               2               3</span></span>
<span><span class="co">#&gt; GGAGACGATTCGTT-1              13              13</span></span>
<span><span class="co">#&gt; CACCGTTGTCGTAG-1              14              14</span></span>
<span><span class="co">#&gt; TATCGTACACGCAT-1               8               8</span></span>
<span><span class="co">#&gt; TACGAGACCTATTC-1               6               1</span></span>
<span><span class="co">#&gt; GTACTACTCATACG-1              14              14</span></span>
<span><span class="co">#&gt;                  RNA_snn_res.1.8 RNA_snn_res.1.9</span></span>
<span><span class="co">#&gt; AGGGCGCTATTTCC-1               1               2</span></span>
<span><span class="co">#&gt; GGAGACGATTCGTT-1              14              14</span></span>
<span><span class="co">#&gt; CACCGTTGTCGTAG-1              15              15</span></span>
<span><span class="co">#&gt; TATCGTACACGCAT-1               8               8</span></span>
<span><span class="co">#&gt; TACGAGACCTATTC-1               3               1</span></span>
<span><span class="co">#&gt; GTACTACTCATACG-1              15              15</span></span>
<span><span class="co">#&gt;                  RNA_snn_res.2 Naive_CD4_T1  cell_label</span></span>
<span><span class="co">#&gt; AGGGCGCTATTTCC-1             1   -0.3540023        &lt;NA&gt;</span></span>
<span><span class="co">#&gt; GGAGACGATTCGTT-1            12    2.1376158 Naive_CD4_T</span></span>
<span><span class="co">#&gt; CACCGTTGTCGTAG-1            15   -1.1487836        &lt;NA&gt;</span></span>
<span><span class="co">#&gt; TATCGTACACGCAT-1             8    0.5557941        &lt;NA&gt;</span></span>
<span><span class="co">#&gt; TACGAGACCTATTC-1             7    1.4250065 Naive_CD4_T</span></span>
<span><span class="co">#&gt; GTACTACTCATACG-1            15   -0.2082793        &lt;NA&gt;</span></span>
<span><span class="co">#&gt;                  SingleR.labels     S.Score   G2M.Score</span></span>
<span><span class="co">#&gt; AGGGCGCTATTTCC-1       Monocyte -0.06808986 -0.08878508</span></span>
<span><span class="co">#&gt; GGAGACGATTCGTT-1        T_cells  0.09975063 -0.06812238</span></span>
<span><span class="co">#&gt; CACCGTTGTCGTAG-1       Monocyte -0.12770457  0.16052598</span></span>
<span><span class="co">#&gt; TATCGTACACGCAT-1    Neutrophils  0.08685571 -0.08316351</span></span>
<span><span class="co">#&gt; TACGAGACCTATTC-1        T_cells -0.02572695  0.02074759</span></span>
<span><span class="co">#&gt; GTACTACTCATACG-1       Monocyte -0.07294007 -0.01517046</span></span>
<span><span class="co">#&gt;                  Phase</span></span>
<span><span class="co">#&gt; AGGGCGCTATTTCC-1    G1</span></span>
<span><span class="co">#&gt; GGAGACGATTCGTT-1     S</span></span>
<span><span class="co">#&gt; CACCGTTGTCGTAG-1   G2M</span></span>
<span><span class="co">#&gt; TATCGTACACGCAT-1     S</span></span>
<span><span class="co">#&gt; TACGAGACCTATTC-1   G2M</span></span>
<span><span class="co">#&gt; GTACTACTCATACG-1    G1</span></span></code></pre></div>
<p>We can then check the cell phase on the UMAP. In this dataset, phase isn’t driving the clustering, and would not require any further handling.</p>
<div class="sourceCode" id="cb101"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://satijalab.org/seurat/reference/DimPlot.html">DimPlot</a></span><span class="op">(</span><span class="va">seurat_object</span>, reduction <span class="op">=</span> <span class="st">'umap_harmony'</span>, group.by <span class="op">=</span> <span class="st">"Phase"</span><span class="op">)</span></span></code></pre></div>
<div class="inline-figure"><img src="scRNAseqInR_ABACBS_2024_Doco_files/figure-html/unnamed-chunk-63-1.png" width="672"></div>
<p>Where a bias <em>is</em> present, your course of action depends on the task at hand. It might involve ‘regressing out’ the cell cycle variation when scaling data <code>ScaleData(kang, vars.to.regress="Phase")</code>, omitting cell-cycle dominated clusters, or just accounting for it in your differential expression calculations.</p>
<p>If you are working with non-human data, you will need to convert these gene lists, or find new cell cycle associated genes in your species.</p>

</div>



  <div class="chapter-nav">
<div class="prev"><a href="de2.html"><span class="header-section-number">13</span> Differential Expression</a></div>
<div class="next"><a href="resources.html"><span class="header-section-number">15</span> Resources</a></div>
</div></main><div class="col-md-3 col-lg-2 d-none d-md-block sidebar sidebar-chapter">
    <nav id="toc" data-toggle="toc" aria-label="On this page"><h2>On this page</h2>
      <ul class="nav navbar-nav"><li><a class="nav-link" href="#CellCycle"><span class="header-section-number">14</span> Cell cycle Assignment</a></li></ul>

      <div class="book-extra">
        <ul class="list-unstyled">
<li><a id="book-source" href="https://monashbioinformaticsplatform.github.io/scRNAseq_Workshop_ABACBS_2024//blob/master/04-11-CellCycle.Rmd">View source <i class="fab fa-gitlab"></i></a></li>
          <li><a id="book-edit" href="https://monashbioinformaticsplatform.github.io/scRNAseq_Workshop_ABACBS_2024//edit/master/04-11-CellCycle.Rmd">Edit this page <i class="fab fa-gitlab"></i></a></li>
        </ul>
</div>
    </nav>
</div>

</div>
</div> <!-- .container -->

<footer class="bg-primary text-light mt-5"><div class="container"><div class="row">

  <div class="col-12 col-md-6 mt-3">
    <p>"<strong>scRNAseq Analysis in R with Seurat</strong>" was written by Monash Genomics and Bioinformatics Platform (MGBP). It was last built on Compiled: November 04, 2024.</p>
  </div>

  <div class="col-12 col-md-6 mt-3">
    <p>This book was built by the <a class="text-light" href="https://bookdown.org">bookdown</a> R package.</p>
  </div>

</div></div>
</footer>
</body>
</html>
