<!DOCTYPE html>
<html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Chapter 11 Cluster Markers | scRNAseq Analysis in R with Seurat</title>
<meta name="author" content="Monash Genomics and Bioinformatics Platform (MGBP)">
<meta name="description" content="Why do we need to do this? Single cell data helps to segragate cell types. Use markers to identify cell types. warning: In this example the cell types/markers are well known and making this step...">
<meta name="generator" content="bookdown 0.41 with bs4_book()">
<meta property="og:title" content="Chapter 11 Cluster Markers | scRNAseq Analysis in R with Seurat">
<meta property="og:type" content="book">
<meta property="og:description" content="Why do we need to do this? Single cell data helps to segragate cell types. Use markers to identify cell types. warning: In this example the cell types/markers are well known and making this step...">
<meta name="twitter:card" content="summary">
<meta name="twitter:title" content="Chapter 11 Cluster Markers | scRNAseq Analysis in R with Seurat">
<meta name="twitter:description" content="Why do we need to do this? Single cell data helps to segragate cell types. Use markers to identify cell types. warning: In this example the cell types/markers are well known and making this step...">
<!-- JS --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://kit.fontawesome.com/6ecbd6c532.js" crossorigin="anonymous"></script><script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="libs/bootstrap-4.6.0/bootstrap.min.css" rel="stylesheet">
<script src="libs/bootstrap-4.6.0/bootstrap.bundle.min.js"></script><script src="libs/bs3compat-0.8.0/transition.js"></script><script src="libs/bs3compat-0.8.0/tabs.js"></script><script src="libs/bs3compat-0.8.0/bs3compat.js"></script><link href="libs/bs4_book-1.0.0/bs4_book.css" rel="stylesheet">
<script src="libs/bs4_book-1.0.0/bs4_book.js"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- CSS --><style type="text/css">
    
    div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  </style>
<link rel="stylesheet" href="style.css">
</head>
<body data-spy="scroll" data-target="#toc">

<div class="container-fluid">
<div class="row">
  <header class="col-sm-12 col-lg-3 sidebar sidebar-book"><a class="sr-only sr-only-focusable" href="#content">Skip to main content</a>

    <div class="d-flex align-items-start justify-content-between">
      <h1>
        <a href="index.html" title="">scRNAseq Analysis in R with Seurat</a>
      </h1>
      <button class="btn btn-outline-primary d-lg-none ml-2 mt-1" type="button" data-toggle="collapse" data-target="#main-nav" aria-expanded="true" aria-controls="main-nav"><i class="fas fa-bars"></i><span class="sr-only">Show table of contents</span></button>
    </div>

    <div id="main-nav" class="collapse-lg">
      <form role="search">
        <input id="search" class="form-control" type="search" placeholder="Search" aria-label="Search">
</form>

      <nav aria-label="Table of contents"><h2>Table of contents</h2>
        <ul class="book-toc list-unstyled">
<li><a class="" href="index.html"><span class="header-section-number">1</span> Getting started</a></li>
<li><a class="" href="set-up.html"><span class="header-section-number">2</span> Set up</a></li>
<li><a class="" href="schedule.html"><span class="header-section-number">3</span> Schedule</a></li>
<li><a class="" href="load.html"><span class="header-section-number">4</span> Load data</a></li>
<li class="book-part">Single Cell Analysis</li>
<li><a class="" href="qc.html"><span class="header-section-number">5</span> QC Filtering</a></li>
<li><a class="" href="norm.html"><span class="header-section-number">6</span> Normalisation</a></li>
<li><a class="" href="reducedims.html"><span class="header-section-number">7</span> PCAs and UMAPs</a></li>
<li><a class="" href="dimensionality-reduction.html"><span class="header-section-number">8</span> Dimensionality reduction</a></li>
<li><a class="" href="Harmony.html"><span class="header-section-number">9</span> Data set integration with Harmony</a></li>
<li><a class="" href="clustering.html"><span class="header-section-number">10</span> Clustering</a></li>
<li><a class="active" href="clustermarkers.html"><span class="header-section-number">11</span> Cluster Markers</a></li>
<li class="book-part">Futher Analysis</li>
<li><a class="" href="singler.html"><span class="header-section-number">12</span> SingleR</a></li>
<li><a class="" href="de2.html"><span class="header-section-number">13</span> Differential Expression</a></li>
<li><a class="" href="CellCycle.html"><span class="header-section-number">14</span> Cell cycle Assignment</a></li>
<li class="book-part">Other Resources</li>
<li><a class="" href="resources.html"><span class="header-section-number">15</span> Resources</a></li>
<li class="book-part">Seurat Object</li>
<li><a class="" href="structure.html"><span class="header-section-number">16</span> Structure</a></li>
<li><a class="" href="acknowledgements.html"><span class="header-section-number">17</span> Acknowledgements</a></li>
<li><a class="" href="session-info.html"><span class="header-section-number">18</span> Session info</a></li>
</ul>

        <div class="book-extra">
          <p><a id="book-repo" href="https://monashbioinformaticsplatform.github.io/scRNAseq_Workshop_ABACBS_2024/">View book source <i class="fab fa-gitlab"></i></a></p>
        </div>
      </nav>
</div>
  </header><main class="col-sm-12 col-md-9 col-lg-7" id="content"><div id="clustermarkers" class="section level1" number="11">
<h1>
<span class="header-section-number">11</span> Cluster Markers<a class="anchor" aria-label="anchor" href="#clustermarkers"><i class="fas fa-link"></i></a>
</h1>
<div id="why-do-we-need-to-do-this-7" class="section level4 unnumbered rational">
<h4>Why do we need to do this?<a class="anchor" aria-label="anchor" href="#why-do-we-need-to-do-this-7"><i class="fas fa-link"></i></a>
</h4>
<p>Single cell data helps to segragate cell types. Use markers to identify cell types. warning: In this example the cell types/markers are well known and making this step easy, but in reality this step needs the experts curation.</p>
</div>
<div id="section-12" class="section level4 unnumbered">
<h4 class="unnumbered"><a class="anchor" aria-label="anchor" href="#section-12"><i class="fas fa-link"></i></a></h4>
</div>
<div id="finding-differentially-expressed-features-cluster-biomarkers" class="section level2" number="11.1">
<h2>
<span class="header-section-number">11.1</span> Finding differentially expressed features (cluster biomarkers)<a class="anchor" aria-label="anchor" href="#finding-differentially-expressed-features-cluster-biomarkers"><i class="fas fa-link"></i></a>
</h2>
<p>Seurat can help you find markers that define clusters via differential expression. By default, it identifies positive and negative markers of a single cluster (specified in <code>ident.1</code>), compared to all other cells. <code><a href="https://satijalab.org/seurat/reference/FindAllMarkers.html">FindAllMarkers()</a></code> automates this process for all clusters, but you can also test groups of clusters vs. each other, or against all cells.</p>
<p>The <code>min.pct</code> argument requires a feature to be detected at a minimum percentage in either of the two groups of cells, and the thresh.test argument requires a feature to be differentially expressed (on average) by some amount between the two groups. You can set both of these to 0, but with a dramatic increase in time - since this will test a large number of features that are unlikely to be highly discriminatory. As another option to speed up these computations, <code>max.cells.per.ident</code> can be set. This will downsample each identity class to have no more cells than whatever this is set to. While there is generally going to be a loss in power, the speed increases can be significant and the most highly differentially expressed features will likely still rise to the top.</p>
<div class="sourceCode" id="cb54"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># find all markers of cluster 2</span></span>
<span><span class="va">cluster2.markers</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/FindMarkers.html">FindMarkers</a></span><span class="op">(</span><span class="va">seurat_object</span>, ident.1 <span class="op">=</span> <span class="fl">2</span>, min.pct <span class="op">=</span> <span class="fl">0.25</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">cluster2.markers</span>, n <span class="op">=</span> <span class="fl">5</span><span class="op">)</span></span>
<span><span class="co">#&gt;      p_val avg_log2FC pct.1 pct.2 p_val_adj</span></span>
<span><span class="co">#&gt; NKG7     0   5.661923 0.800 0.056         0</span></span>
<span><span class="co">#&gt; CCL5     0   4.482005 0.749 0.107         0</span></span>
<span><span class="co">#&gt; GZMB     0   5.281761 0.652 0.037         0</span></span>
<span><span class="co">#&gt; GNLY     0   6.095800 0.661 0.049         0</span></span>
<span><span class="co">#&gt; CST7     0   4.799111 0.627 0.044         0</span></span>
<span><span class="co"># find all markers distinguishing cluster 5 from clusters 0 and 3</span></span>
<span><span class="va">cluster5.markers</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/FindMarkers.html">FindMarkers</a></span><span class="op">(</span><span class="va">seurat_object</span>, ident.1 <span class="op">=</span> <span class="fl">5</span>, ident.2 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">3</span><span class="op">)</span>, min.pct <span class="op">=</span> <span class="fl">0.25</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">cluster5.markers</span>, n <span class="op">=</span> <span class="fl">5</span><span class="op">)</span></span>
<span><span class="co">#&gt;                p_val avg_log2FC pct.1 pct.2     p_val_adj</span></span>
<span><span class="co">#&gt; HSPH1  2.191916e-116   4.608521 0.591 0.060 7.810891e-112</span></span>
<span><span class="co">#&gt; KLF10   7.713460e-63   4.173941 0.274 0.020  2.748692e-58</span></span>
<span><span class="co">#&gt; HSPA1A  1.488828e-62   5.250111 0.378 0.046  5.305439e-58</span></span>
<span><span class="co">#&gt; SRSF2   1.282249e-60   2.836025 0.689 0.184  4.569294e-56</span></span>
<span><span class="co">#&gt; SNHG15  9.077911e-60   3.667237 0.390 0.051  3.234914e-55</span></span>
<span><span class="co"># find markers for every cluster compared to all remaining cells, report only the positive ones</span></span>
<span><span class="va">seurat_object.markers</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/FindAllMarkers.html">FindAllMarkers</a></span><span class="op">(</span><span class="va">seurat_object</span>, only.pos <span class="op">=</span> <span class="cn">TRUE</span>, min.pct <span class="op">=</span> <span class="fl">0.25</span>, logfc.threshold <span class="op">=</span> <span class="fl">0.25</span><span class="op">)</span></span>
<span><span class="co">#&gt; Calculating cluster 0</span></span>
<span><span class="co">#&gt; Calculating cluster 1</span></span>
<span><span class="co">#&gt; Calculating cluster 2</span></span>
<span><span class="co">#&gt; Calculating cluster 3</span></span>
<span><span class="co">#&gt; Calculating cluster 4</span></span>
<span><span class="co">#&gt; Calculating cluster 5</span></span>
<span><span class="co">#&gt; Calculating cluster 6</span></span>
<span><span class="co">#&gt; Calculating cluster 7</span></span>
<span><span class="va">seurat_object.markers</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html">group_by</a></span><span class="op">(</span><span class="va">cluster</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/slice.html">slice_max</a></span><span class="op">(</span>n <span class="op">=</span> <span class="fl">2</span>, order_by <span class="op">=</span> <span class="va">avg_log2FC</span><span class="op">)</span></span>
<span><span class="co">#&gt; # A tibble: 16 × 7</span></span>
<span><span class="co">#&gt; # Groups:   cluster [8]</span></span>
<span><span class="co">#&gt;        p_val avg_log2FC pct.1 pct.2 p_val_adj cluster gene  </span></span>
<span><span class="co">#&gt;        &lt;dbl&gt;      &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt;     &lt;dbl&gt; &lt;fct&gt;   &lt;chr&gt; </span></span>
<span><span class="co">#&gt;  1 8.51e-196       2.11 0.614 0.223 3.03e-191 0       SELL  </span></span>
<span><span class="co">#&gt;  2 3.26e-223       2.05 0.653 0.221 1.16e-218 0       LTB   </span></span>
<span><span class="co">#&gt;  3 0               7.02 0.672 0.021 0         1       S100A8</span></span>
<span><span class="co">#&gt;  4 2.19e-267       7.00 0.353 0.013 7.80e-263 1       CCL7  </span></span>
<span><span class="co">#&gt;  5 0               6.10 0.661 0.049 0         2       GNLY  </span></span>
<span><span class="co">#&gt;  6 4.38e-255       6.07 0.331 0.011 1.56e-250 2       GZMA  </span></span>
<span><span class="co">#&gt;  7 0               6.29 0.442 0.011 0         3       MS4A1 </span></span>
<span><span class="co">#&gt;  8 0               6.17 0.657 0.018 0         3       CD79A </span></span>
<span><span class="co">#&gt;  9 0               6.75 0.833 0.029 0         4       VMO1  </span></span>
<span><span class="co">#&gt; 10 0               5.19 0.742 0.036 0         4       MS4A4A</span></span>
<span><span class="co">#&gt; 11 2.18e-121       4.63 0.591 0.076 7.76e-117 5       HSPH1 </span></span>
<span><span class="co">#&gt; 12 3.46e- 90       4.21 0.39  0.04  1.23e- 85 5       SNHG15</span></span>
<span><span class="co">#&gt; 13 2.72e-243       7.17 0.438 0.011 9.68e-239 6       PF4   </span></span>
<span><span class="co">#&gt; 14 1.51e-120       6.57 0.253 0.009 5.38e-116 6       SDPR  </span></span>
<span><span class="co">#&gt; 15 0               8.79 0.438 0.001 0         7       CCL22 </span></span>
<span><span class="co">#&gt; 16 0               7.42 0.517 0.003 0         7       PKIB</span></span></code></pre></div>
<p>Seurat has several tests for differential expression which can be set with the test.use parameter (see our <a href="de_vignette.html">DE vignette</a> for details). For example, the ROC test returns the ‘classification power’ <code>abs(AUC-0.5)*2</code> for any individual marker, ranging from 0 = random to 1 = perfect.</p>
<div class="sourceCode" id="cb55"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">cluster0.markers</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/FindMarkers.html">FindMarkers</a></span><span class="op">(</span><span class="va">seurat_object</span>, ident.1 <span class="op">=</span> <span class="fl">0</span>, logfc.threshold <span class="op">=</span> <span class="fl">0.25</span>, test.use <span class="op">=</span> <span class="st">"roc"</span>, only.pos <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p>We include several tools for visualizing marker expression. <code><a href="https://satijalab.org/seurat/reference/VlnPlot.html">VlnPlot()</a></code> (shows expression probability distributions across clusters), and <code><a href="https://satijalab.org/seurat/reference/FeaturePlot.html">FeaturePlot()</a></code> (visualizes feature expression on a tSNE or PCA plot) are our most commonly used visualizations. We also suggest exploring <code><a href="https://satijalab.org/seurat/reference/RidgePlot.html">RidgePlot()</a></code>, <code><a href="https://satijalab.org/seurat/reference/CellScatter.html">CellScatter()</a></code>, and <code><a href="https://satijalab.org/seurat/reference/DotPlot.html">DotPlot()</a></code> as additional methods to view your dataset.</p>
<div class="sourceCode" id="cb56"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://satijalab.org/seurat/reference/VlnPlot.html">VlnPlot</a></span><span class="op">(</span><span class="va">seurat_object</span>, features <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"MS4A1"</span>, <span class="st">"CD79A"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<div class="inline-figure"><img src="scRNAseqInR_ABACBS_2024_Doco_files/figure-html/unnamed-chunk-28-1.png" width="672"></div>
<div class="sourceCode" id="cb57"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># you can plot raw counts as well</span></span>
<span><span class="fu"><a href="https://satijalab.org/seurat/reference/VlnPlot.html">VlnPlot</a></span><span class="op">(</span><span class="va">seurat_object</span>, features <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"NKG7"</span>, <span class="st">"PF4"</span><span class="op">)</span>, slot <span class="op">=</span> <span class="st">'counts'</span>, log <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="co">#&gt; Warning: The `slot` argument of `VlnPlot()` is deprecated as of Seurat 5.0.0.</span></span>
<span><span class="co">#&gt; ℹ Please use the `layer` argument instead.</span></span>
<span><span class="co">#&gt; This warning is displayed once every 8 hours.</span></span>
<span><span class="co">#&gt; Call `lifecycle::last_lifecycle_warnings()` to see where this warning was generated.</span></span></code></pre></div>
<div class="inline-figure"><img src="scRNAseqInR_ABACBS_2024_Doco_files/figure-html/unnamed-chunk-28-2.png" width="672"></div>
<div class="sourceCode" id="cb58"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://satijalab.org/seurat/reference/FeaturePlot.html">FeaturePlot</a></span><span class="op">(</span><span class="va">seurat_object</span>, features <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"MS4A1"</span>, <span class="st">"GNLY"</span>, <span class="st">"CD3E"</span>, <span class="st">"CD14"</span>, <span class="st">"FCER1A"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<div class="inline-figure"><img src="scRNAseqInR_ABACBS_2024_Doco_files/figure-html/markerplots-1.png" width="672"></div>
<div class="sourceCode" id="cb59"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://satijalab.org/seurat/reference/FeaturePlot.html">FeaturePlot</a></span><span class="op">(</span><span class="va">seurat_object</span>, features <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"FCGR3A"</span>, <span class="st">"LYZ"</span>, <span class="st">"PPBP"</span>, <span class="st">"CD8A"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<div class="inline-figure"><img src="scRNAseqInR_ABACBS_2024_Doco_files/figure-html/markerplots-2.png" width="672"></div>
<details><summary><strong>Other useful plots</strong>
</summary><p>These are ridgeplots, cell scatter plots and dotplots. Replace <code>FeaturePlot</code> with the other functions.</p>
<div class="sourceCode" id="cb60"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://satijalab.org/seurat/reference/RidgePlot.html">RidgePlot</a></span><span class="op">(</span><span class="va">seurat_object</span>, features <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"MS4A1"</span>, <span class="st">"GNLY"</span>, <span class="st">"CD3E"</span>, <span class="st">"CD14"</span>, <span class="st">"FCER1A"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; Picking joint bandwidth of 0.0426</span></span>
<span><span class="co">#&gt; Picking joint bandwidth of 0.0658</span></span>
<span><span class="co">#&gt; Picking joint bandwidth of 0.0723</span></span>
<span><span class="co">#&gt; Picking joint bandwidth of 0.0901</span></span>
<span><span class="co">#&gt; Picking joint bandwidth of 2.99e-06</span></span></code></pre></div>
<div class="inline-figure"><img src="scRNAseqInR_ABACBS_2024_Doco_files/figure-html/ridgeplots-1.png" width="672"></div>
<div class="sourceCode" id="cb61"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://satijalab.org/seurat/reference/RidgePlot.html">RidgePlot</a></span><span class="op">(</span><span class="va">seurat_object</span>, features <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"FCGR3A"</span>, <span class="st">"LYZ"</span>, <span class="st">"PPBP"</span>, <span class="st">"CD8A"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; Picking joint bandwidth of 0.0546</span></span>
<span><span class="co">#&gt; Picking joint bandwidth of 0.136</span></span>
<span><span class="co">#&gt; Picking joint bandwidth of 0.0843</span></span>
<span><span class="co">#&gt; Picking joint bandwidth of 0.0343</span></span></code></pre></div>
<p><img src="scRNAseqInR_ABACBS_2024_Doco_files/figure-html/ridgeplots-2.png" width="672">
For CellScatter plots, will need the cell id of the cells you want to look at. You can get this from the cell metadata (<code>seurat_object@meta.data</code>).</p>
<div class="sourceCode" id="cb62"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span> <span class="va">seurat_object</span><span class="op">@</span><span class="va">meta.data</span> <span class="op">)</span></span>
<span><span class="co">#&gt;                     orig.ident nCount_RNA nFeature_RNA  ind</span></span>
<span><span class="co">#&gt; AGGGCGCTATTTCC-1 SeuratProject       2053          532 1256</span></span>
<span><span class="co">#&gt; GGAGACGATTCGTT-1 SeuratProject        881          392 1256</span></span>
<span><span class="co">#&gt; CACCGTTGTCGTAG-1 SeuratProject       3130         1005 1016</span></span>
<span><span class="co">#&gt; TATCGTACACGCAT-1 SeuratProject       1042          549 1488</span></span>
<span><span class="co">#&gt; TACGAGACCTATTC-1 SeuratProject       2425          777 1244</span></span>
<span><span class="co">#&gt; GTACTACTCATACG-1 SeuratProject       3951         1064 1256</span></span>
<span><span class="co">#&gt;                  stim              cell multiplets</span></span>
<span><span class="co">#&gt; AGGGCGCTATTTCC-1 stim   CD14+ Monocytes    singlet</span></span>
<span><span class="co">#&gt; GGAGACGATTCGTT-1 stim       CD4 T cells    singlet</span></span>
<span><span class="co">#&gt; CACCGTTGTCGTAG-1 ctrl FCGR3A+ Monocytes    singlet</span></span>
<span><span class="co">#&gt; TATCGTACACGCAT-1 stim           B cells    singlet</span></span>
<span><span class="co">#&gt; TACGAGACCTATTC-1 stim       CD4 T cells    singlet</span></span>
<span><span class="co">#&gt; GTACTACTCATACG-1 ctrl FCGR3A+ Monocytes    singlet</span></span>
<span><span class="co">#&gt;                  percent.mt RNA_snn_res.0.5 seurat_clusters</span></span>
<span><span class="co">#&gt; AGGGCGCTATTTCC-1  1.6336634               1               0</span></span>
<span><span class="co">#&gt; GGAGACGATTCGTT-1  4.8809524               6              15</span></span>
<span><span class="co">#&gt; CACCGTTGTCGTAG-1  1.0655473               4              11</span></span>
<span><span class="co">#&gt; TATCGTACACGCAT-1  3.0662710               3               4</span></span>
<span><span class="co">#&gt; TACGAGACCTATTC-1  1.0837849               0               8</span></span>
<span><span class="co">#&gt; GTACTACTCATACG-1  0.7137395               4              11</span></span>
<span><span class="co">#&gt;                  pca_clusters harmony_clusters</span></span>
<span><span class="co">#&gt; AGGGCGCTATTTCC-1            3                1</span></span>
<span><span class="co">#&gt; GGAGACGATTCGTT-1            0                6</span></span>
<span><span class="co">#&gt; CACCGTTGTCGTAG-1            9                4</span></span>
<span><span class="co">#&gt; TATCGTACACGCAT-1            6                3</span></span>
<span><span class="co">#&gt; TACGAGACCTATTC-1            0                0</span></span>
<span><span class="co">#&gt; GTACTACTCATACG-1            9                4</span></span>
<span><span class="co">#&gt;                  RNA_snn_res.0.1 RNA_snn_res.0.2</span></span>
<span><span class="co">#&gt; AGGGCGCTATTTCC-1               1               1</span></span>
<span><span class="co">#&gt; GGAGACGATTCGTT-1               0               0</span></span>
<span><span class="co">#&gt; CACCGTTGTCGTAG-1               4               4</span></span>
<span><span class="co">#&gt; TATCGTACACGCAT-1               3               3</span></span>
<span><span class="co">#&gt; TACGAGACCTATTC-1               0               0</span></span>
<span><span class="co">#&gt; GTACTACTCATACG-1               4               4</span></span>
<span><span class="co">#&gt;                  RNA_snn_res.0.3 RNA_snn_res.0.4</span></span>
<span><span class="co">#&gt; AGGGCGCTATTTCC-1               1               1</span></span>
<span><span class="co">#&gt; GGAGACGATTCGTT-1               0               6</span></span>
<span><span class="co">#&gt; CACCGTTGTCGTAG-1               4               4</span></span>
<span><span class="co">#&gt; TATCGTACACGCAT-1               3               3</span></span>
<span><span class="co">#&gt; TACGAGACCTATTC-1               0               0</span></span>
<span><span class="co">#&gt; GTACTACTCATACG-1               4               4</span></span>
<span><span class="co">#&gt;                  RNA_snn_res.0.6 RNA_snn_res.0.7</span></span>
<span><span class="co">#&gt; AGGGCGCTATTTCC-1               1               1</span></span>
<span><span class="co">#&gt; GGAGACGATTCGTT-1               7               7</span></span>
<span><span class="co">#&gt; CACCGTTGTCGTAG-1               5               6</span></span>
<span><span class="co">#&gt; TATCGTACACGCAT-1               3               4</span></span>
<span><span class="co">#&gt; TACGAGACCTATTC-1               0               3</span></span>
<span><span class="co">#&gt; GTACTACTCATACG-1               5               6</span></span>
<span><span class="co">#&gt;                  RNA_snn_res.0.8 RNA_snn_res.0.9</span></span>
<span><span class="co">#&gt; AGGGCGCTATTTCC-1               1               1</span></span>
<span><span class="co">#&gt; GGAGACGATTCGTT-1               9               9</span></span>
<span><span class="co">#&gt; CACCGTTGTCGTAG-1               7               7</span></span>
<span><span class="co">#&gt; TATCGTACACGCAT-1               3               3</span></span>
<span><span class="co">#&gt; TACGAGACCTATTC-1               0               0</span></span>
<span><span class="co">#&gt; GTACTACTCATACG-1               7               7</span></span>
<span><span class="co">#&gt;                  RNA_snn_res.1 RNA_snn_res.1.1</span></span>
<span><span class="co">#&gt; AGGGCGCTATTTCC-1             0               1</span></span>
<span><span class="co">#&gt; GGAGACGATTCGTT-1             9               8</span></span>
<span><span class="co">#&gt; CACCGTTGTCGTAG-1             8               7</span></span>
<span><span class="co">#&gt; TATCGTACACGCAT-1             3               4</span></span>
<span><span class="co">#&gt; TACGAGACCTATTC-1             6               0</span></span>
<span><span class="co">#&gt; GTACTACTCATACG-1             8               7</span></span>
<span><span class="co">#&gt;                  RNA_snn_res.1.2 RNA_snn_res.1.3</span></span>
<span><span class="co">#&gt; AGGGCGCTATTTCC-1               0               2</span></span>
<span><span class="co">#&gt; GGAGACGATTCGTT-1              10              11</span></span>
<span><span class="co">#&gt; CACCGTTGTCGTAG-1               8               9</span></span>
<span><span class="co">#&gt; TATCGTACACGCAT-1               3               3</span></span>
<span><span class="co">#&gt; TACGAGACCTATTC-1               2               7</span></span>
<span><span class="co">#&gt; GTACTACTCATACG-1               8               9</span></span>
<span><span class="co">#&gt;                  RNA_snn_res.1.4 RNA_snn_res.1.5</span></span>
<span><span class="co">#&gt; AGGGCGCTATTTCC-1               0               1</span></span>
<span><span class="co">#&gt; GGAGACGATTCGTT-1              12              12</span></span>
<span><span class="co">#&gt; CACCGTTGTCGTAG-1              10              10</span></span>
<span><span class="co">#&gt; TATCGTACACGCAT-1               1               3</span></span>
<span><span class="co">#&gt; TACGAGACCTATTC-1               2               4</span></span>
<span><span class="co">#&gt; GTACTACTCATACG-1              10              10</span></span>
<span><span class="co">#&gt;                  RNA_snn_res.1.6 RNA_snn_res.1.7</span></span>
<span><span class="co">#&gt; AGGGCGCTATTTCC-1               1               0</span></span>
<span><span class="co">#&gt; GGAGACGATTCGTT-1              12              13</span></span>
<span><span class="co">#&gt; CACCGTTGTCGTAG-1              10              11</span></span>
<span><span class="co">#&gt; TATCGTACACGCAT-1               2               1</span></span>
<span><span class="co">#&gt; TACGAGACCTATTC-1               9               7</span></span>
<span><span class="co">#&gt; GTACTACTCATACG-1              10              11</span></span>
<span><span class="co">#&gt;                  RNA_snn_res.1.8 RNA_snn_res.1.9</span></span>
<span><span class="co">#&gt; AGGGCGCTATTTCC-1               0               7</span></span>
<span><span class="co">#&gt; GGAGACGATTCGTT-1              13              14</span></span>
<span><span class="co">#&gt; CACCGTTGTCGTAG-1              11              12</span></span>
<span><span class="co">#&gt; TATCGTACACGCAT-1               1               0</span></span>
<span><span class="co">#&gt; TACGAGACCTATTC-1               8               8</span></span>
<span><span class="co">#&gt; GTACTACTCATACG-1              11              12</span></span>
<span><span class="co">#&gt;                  RNA_snn_res.2</span></span>
<span><span class="co">#&gt; AGGGCGCTATTTCC-1             0</span></span>
<span><span class="co">#&gt; GGAGACGATTCGTT-1            15</span></span>
<span><span class="co">#&gt; CACCGTTGTCGTAG-1            11</span></span>
<span><span class="co">#&gt; TATCGTACACGCAT-1             4</span></span>
<span><span class="co">#&gt; TACGAGACCTATTC-1             8</span></span>
<span><span class="co">#&gt; GTACTACTCATACG-1            11</span></span>
<span></span>
<span><span class="fu"><a href="https://satijalab.org/seurat/reference/CellScatter.html">CellScatter</a></span><span class="op">(</span><span class="va">seurat_object</span>, cell1 <span class="op">=</span> <span class="st">"AGGGCGCTATTTCC-1"</span>, cell2 <span class="op">=</span> <span class="st">"GGAGACGATTCGTT-1"</span><span class="op">)</span></span></code></pre></div>
<div class="inline-figure"><img src="scRNAseqInR_ABACBS_2024_Doco_files/figure-html/unnamed-chunk-29-1.png" width="672"></div>
<div class="sourceCode" id="cb63"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="fu"><a href="https://satijalab.org/seurat/reference/CellScatter.html">CellScatter</a></span><span class="op">(</span><span class="va">seurat_object</span>, cell1 <span class="op">=</span> <span class="st">"GGAGACGATTCGTT-1"</span>, cell2 <span class="op">=</span> <span class="st">"TACGAGACCTATTC-1"</span><span class="op">)</span></span></code></pre></div>
<div class="inline-figure"><img src="scRNAseqInR_ABACBS_2024_Doco_files/figure-html/unnamed-chunk-29-2.png" width="672"></div>
<p>DotPlots</p>
<div class="sourceCode" id="cb64"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://satijalab.org/seurat/reference/DotPlot.html">DotPlot</a></span><span class="op">(</span><span class="va">seurat_object</span>, features <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"MS4A1"</span>, <span class="st">"GNLY"</span>, <span class="st">"CD3E"</span>, <span class="st">"CD14"</span>, <span class="st">"FCER1A"</span>, <span class="st">"FCGR3A"</span>, <span class="st">"LYZ"</span>, <span class="st">"PPBP"</span>, <span class="st">"CD8A"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<div class="inline-figure"><img src="scRNAseqInR_ABACBS_2024_Doco_files/figure-html/unnamed-chunk-30-1.png" width="672"></div>
</details><p><code><a href="https://satijalab.org/seurat/reference/DoHeatmap.html">DoHeatmap()</a></code> generates an expression heatmap for given cells and features. In this case, we are plotting the top 10 markers (or all markers if less than 10) for each cluster.</p>
<div class="sourceCode" id="cb65"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">maxcells</span><span class="op">=</span><span class="fl">1500</span></span>
<span><span class="va">top10</span> <span class="op">&lt;-</span> <span class="va">seurat_object.markers</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html">group_by</a></span><span class="op">(</span><span class="va">cluster</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/top_n.html">top_n</a></span><span class="op">(</span>n <span class="op">=</span> <span class="fl">10</span>, wt <span class="op">=</span> <span class="va">avg_log2FC</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://satijalab.org/seurat/reference/DoHeatmap.html">DoHeatmap</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/subset.html">subset</a></span><span class="op">(</span><span class="va">seurat_object</span>, downsample <span class="op">=</span> <span class="va">maxcells</span><span class="op">)</span>, features <span class="op">=</span> <span class="va">top10</span><span class="op">$</span><span class="va">gene</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/SeuratTheme.html">NoLegend</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<div class="inline-figure"><img src="scRNAseqInR_ABACBS_2024_Doco_files/figure-html/clusterHeatmap-1.png" width="1440"></div>
</div>
<div id="use-makers-to-label-or-find-a-cluster" class="section level2" number="11.2">
<h2>
<span class="header-section-number">11.2</span> Use makers to label or find a cluster<a class="anchor" aria-label="anchor" href="#use-makers-to-label-or-find-a-cluster"><i class="fas fa-link"></i></a>
</h2>
<p>If you know markers for your cell types, use AddModuleScore to label them.</p>
<div class="sourceCode" id="cb66"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">genes_markers</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span>Naive_CD4_T <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"IL7R"</span>, <span class="st">"CCR7"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">seurat_object</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/AddModuleScore.html">AddModuleScore</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">seurat_object</span>, features <span class="op">=</span> <span class="va">genes_markers</span>, ctrl <span class="op">=</span> <span class="fl">5</span>, name <span class="op">=</span> <span class="st">"Naive_CD4_T"</span>,</span>
<span>    search <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span></span>
<span></span>
<span><span class="co"># notice the name of the cluster has a 1 at the end</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">seurat_object</span><span class="op">@</span><span class="va">meta.data</span><span class="op">)</span></span>
<span><span class="co">#&gt;  [1] "orig.ident"       "nCount_RNA"      </span></span>
<span><span class="co">#&gt;  [3] "nFeature_RNA"     "ind"             </span></span>
<span><span class="co">#&gt;  [5] "stim"             "cell"            </span></span>
<span><span class="co">#&gt;  [7] "multiplets"       "percent.mt"      </span></span>
<span><span class="co">#&gt;  [9] "RNA_snn_res.0.5"  "seurat_clusters" </span></span>
<span><span class="co">#&gt; [11] "pca_clusters"     "harmony_clusters"</span></span>
<span><span class="co">#&gt; [13] "RNA_snn_res.0.1"  "RNA_snn_res.0.2" </span></span>
<span><span class="co">#&gt; [15] "RNA_snn_res.0.3"  "RNA_snn_res.0.4" </span></span>
<span><span class="co">#&gt; [17] "RNA_snn_res.0.6"  "RNA_snn_res.0.7" </span></span>
<span><span class="co">#&gt; [19] "RNA_snn_res.0.8"  "RNA_snn_res.0.9" </span></span>
<span><span class="co">#&gt; [21] "RNA_snn_res.1"    "RNA_snn_res.1.1" </span></span>
<span><span class="co">#&gt; [23] "RNA_snn_res.1.2"  "RNA_snn_res.1.3" </span></span>
<span><span class="co">#&gt; [25] "RNA_snn_res.1.4"  "RNA_snn_res.1.5" </span></span>
<span><span class="co">#&gt; [27] "RNA_snn_res.1.6"  "RNA_snn_res.1.7" </span></span>
<span><span class="co">#&gt; [29] "RNA_snn_res.1.8"  "RNA_snn_res.1.9" </span></span>
<span><span class="co">#&gt; [31] "RNA_snn_res.2"    "Naive_CD4_T1"</span></span>
<span></span>
<span><span class="co"># label that cell type</span></span>
<span><span class="va">seurat_object</span><span class="op">$</span><span class="va">cell_label</span> <span class="op">=</span> <span class="cn">NA</span></span>
<span><span class="va">seurat_object</span><span class="op">$</span><span class="va">cell_label</span><span class="op">[</span><span class="va">seurat_object</span><span class="op">$</span><span class="va">Naive_CD4_T1</span> <span class="op">&gt;</span> <span class="fl">1</span><span class="op">]</span> <span class="op">=</span> <span class="st">"Naive_CD4_T"</span></span>
<span><span class="fu"><a href="https://mojaveazure.github.io/seurat-object/reference/Idents.html">Idents</a></span><span class="op">(</span><span class="va">seurat_object</span><span class="op">)</span> <span class="op">=</span> <span class="va">seurat_object</span><span class="op">$</span><span class="va">cell_label</span></span>
<span></span>
<span><span class="co"># plot</span></span>
<span><span class="co"># Using a custom colour scale </span></span>
<span><span class="fu"><a href="https://satijalab.org/seurat/reference/FeaturePlot.html">FeaturePlot</a></span><span class="op">(</span><span class="va">seurat_object</span>, features <span class="op">=</span> <span class="st">"Naive_CD4_T1"</span>, label <span class="op">=</span> <span class="cn">TRUE</span>, repel <span class="op">=</span> <span class="cn">TRUE</span>, reduction <span class="op">=</span> <span class="st">"umap_harmony"</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_gradient.html">scale_colour_gradientn</a></span><span class="op">(</span>colours <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"lightblue"</span>,<span class="st">"beige"</span>,<span class="st">"red"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; Scale for colour is already present.</span></span>
<span><span class="co">#&gt; Adding another scale for colour, which will replace the existing scale.</span></span></code></pre></div>
<div class="inline-figure"><img src="scRNAseqInR_ABACBS_2024_Doco_files/figure-html/unnamed-chunk-31-1.png" width="672"></div>
</div>
<div id="assigning-cell-type-identity-to-clusters" class="section level2" number="11.3">
<h2>
<span class="header-section-number">11.3</span> Assigning cell type identity to clusters<a class="anchor" aria-label="anchor" href="#assigning-cell-type-identity-to-clusters"><i class="fas fa-link"></i></a>
</h2>
<p>Fortunately in the case of this dataset, we can use canonical markers to easily match the unbiased clustering to known cell types:</p>
<div class="inline-table"><table class="table table-sm">
<thead><tr class="header">
<th>Markers</th>
<th>Cell Type</th>
</tr></thead>
<tbody>
<tr class="odd">
<td>IL7R, CCR7</td>
<td>Naive CD4+ T</td>
</tr>
<tr class="even">
<td>CD14, LYZ</td>
<td>CD14+ Mono</td>
</tr>
<tr class="odd">
<td>IL7R, S100A4</td>
<td>Memory CD4+</td>
</tr>
<tr class="even">
<td>MS4A1</td>
<td>B</td>
</tr>
<tr class="odd">
<td>CD8A</td>
<td>CD8+ T</td>
</tr>
<tr class="even">
<td>FCGR3A, MS4A7</td>
<td>FCGR3A+ Mono</td>
</tr>
<tr class="odd">
<td>GNLY, NKG7</td>
<td>NK</td>
</tr>
<tr class="even">
<td>FCER1A, CST3</td>
<td>DC</td>
</tr>
<tr class="odd">
<td>PPBP</td>
<td>Platelet</td>
</tr>
</tbody>
</table></div>
<div class="sourceCode" id="cb67"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://satijalab.org/seurat/reference/DimPlot.html">DimPlot</a></span><span class="op">(</span><span class="va">seurat_object</span>,group.by <span class="op">=</span> <span class="st">"RNA_snn_res.0.5"</span>,reduction <span class="op">=</span> <span class="st">"umap_harmony"</span><span class="op">)</span><span class="op">|</span><span class="fu"><a href="https://satijalab.org/seurat/reference/FeaturePlot.html">FeaturePlot</a></span><span class="op">(</span><span class="va">seurat_object</span>, features <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span> <span class="st">"MS4A1"</span><span class="op">)</span>,reduction <span class="op">=</span> <span class="st">"umap_harmony"</span><span class="op">)</span></span></code></pre></div>
<div class="inline-figure"><img src="scRNAseqInR_ABACBS_2024_Doco_files/figure-html/labelplot-1.png" width="672"></div>
<div id="challenge-match-cluster-numbers-with-cell-labels" class="section level4 unnumbered challenge">
<h4>Challenge: Match cluster numbers with cell labels<a class="anchor" aria-label="anchor" href="#challenge-match-cluster-numbers-with-cell-labels"><i class="fas fa-link"></i></a>
</h4>
<p>Use the markers provided and the resolution 0.5 to identity the cell labels</p>
<div class="sourceCode" id="cb68"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://mojaveazure.github.io/seurat-object/reference/Idents.html">Idents</a></span><span class="op">(</span><span class="va">seurat_object</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">seurat_object</span><span class="op">$</span><span class="va">RNA_snn_res.0.5</span></span>
<span></span>
<span><span class="co"># This renaming list is incomplete and incorrect.</span></span>
<span><span class="co"># Fix and add to the list to fill in the cell types.</span></span>
<span><span class="va">new.cluster.ids</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span></span>
<span>    <span class="st">"0"</span> <span class="op">=</span> <span class="st">"Naive CD4+ T"</span>,</span>
<span>    <span class="st">"2"</span> <span class="op">=</span> <span class="st">"CD14+ Monocytes"</span>,</span>
<span>    <span class="st">"1"</span> <span class="op">=</span> <span class="st">"B cells"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">seurat_object</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://mojaveazure.github.io/seurat-object/reference/Idents.html">RenameIdents</a></span><span class="op">(</span><span class="va">seurat_object</span>, <span class="va">new.cluster.ids</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://satijalab.org/seurat/reference/DimPlot.html">DimPlot</a></span><span class="op">(</span><span class="va">seurat_object</span>, reduction <span class="op">=</span> <span class="st">'umap_harmony'</span>, label <span class="op">=</span> <span class="cn">TRUE</span>, pt.size <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/SeuratTheme.html">NoLegend</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
</div>
<div id="section-13" class="section level4 unnumbered">
<h4 class="unnumbered"><a class="anchor" aria-label="anchor" href="#section-13"><i class="fas fa-link"></i></a></h4>
</div>
<div id="save-the-plots" class="section level4" number="11.3.0.1">
<h4>
<span class="header-section-number">11.3.0.1</span> save the plots<a class="anchor" aria-label="anchor" href="#save-the-plots"><i class="fas fa-link"></i></a>
</h4>
<div class="sourceCode" id="cb69"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="va">plot</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/DimPlot.html">DimPlot</a></span><span class="op">(</span><span class="va">seurat_object</span>, reduction <span class="op">=</span> <span class="st">'umap_harmony'</span>, label <span class="op">=</span> <span class="cn">TRUE</span>, label.size <span class="op">=</span> <span class="fl">4.5</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">xlab</a></span><span class="op">(</span><span class="st">"UMAP 1"</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">ylab</a></span><span class="op">(</span><span class="st">"UMAP 2"</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html">theme</a></span><span class="op">(</span>axis.title <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">element_text</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">18</span><span class="op">)</span>, legend.text <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">element_text</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">18</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/guides.html">guides</a></span><span class="op">(</span>colour <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/guide_legend.html">guide_legend</a></span><span class="op">(</span>override.aes <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">10</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggsave.html">ggsave</a></span><span class="op">(</span>filename <span class="op">=</span> <span class="st">"seurat_object3k_umap.jpg"</span>, height <span class="op">=</span> <span class="fl">7</span>, width <span class="op">=</span> <span class="fl">12</span>, plot <span class="op">=</span> <span class="va">plot</span>, quality <span class="op">=</span> <span class="fl">50</span><span class="op">)</span></span></code></pre></div>
</div>
<div id="save-the-seurat-object" class="section level4" number="11.3.0.2">
<h4>
<span class="header-section-number">11.3.0.2</span> save the seurat object<a class="anchor" aria-label="anchor" href="#save-the-seurat-object"><i class="fas fa-link"></i></a>
</h4>
<div class="sourceCode" id="cb70"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/readRDS.html">saveRDS</a></span><span class="op">(</span><span class="va">seurat_object</span>, file <span class="op">=</span> <span class="st">"seurat_object3k_final.rds"</span><span class="op">)</span></span></code></pre></div>

</div>
</div>
</div>



  <div class="chapter-nav">
<div class="prev"><a href="clustering.html"><span class="header-section-number">10</span> Clustering</a></div>
<div class="next"><a href="singler.html"><span class="header-section-number">12</span> SingleR</a></div>
</div></main><div class="col-md-3 col-lg-2 d-none d-md-block sidebar sidebar-chapter">
    <nav id="toc" data-toggle="toc" aria-label="On this page"><h2>On this page</h2>
      <ul class="nav navbar-nav">
<li><a class="nav-link" href="#clustermarkers"><span class="header-section-number">11</span> Cluster Markers</a></li>
<li><a class="nav-link" href="#finding-differentially-expressed-features-cluster-biomarkers"><span class="header-section-number">11.1</span> Finding differentially expressed features (cluster biomarkers)</a></li>
<li><a class="nav-link" href="#use-makers-to-label-or-find-a-cluster"><span class="header-section-number">11.2</span> Use makers to label or find a cluster</a></li>
<li><a class="nav-link" href="#assigning-cell-type-identity-to-clusters"><span class="header-section-number">11.3</span> Assigning cell type identity to clusters</a></li>
</ul>

      <div class="book-extra">
        <ul class="list-unstyled">
<li><a id="book-source" href="https://monashbioinformaticsplatform.github.io/scRNAseq_Workshop_ABACBS_2024//blob/master/04-07-clustermarkers.Rmd">View source <i class="fab fa-gitlab"></i></a></li>
          <li><a id="book-edit" href="https://monashbioinformaticsplatform.github.io/scRNAseq_Workshop_ABACBS_2024//edit/master/04-07-clustermarkers.Rmd">Edit this page <i class="fab fa-gitlab"></i></a></li>
        </ul>
</div>
    </nav>
</div>

</div>
</div> <!-- .container -->

<footer class="bg-primary text-light mt-5"><div class="container"><div class="row">

  <div class="col-12 col-md-6 mt-3">
    <p>"<strong>scRNAseq Analysis in R with Seurat</strong>" was written by Monash Genomics and Bioinformatics Platform (MGBP). It was last built on Compiled: November 11, 2024.</p>
  </div>

  <div class="col-12 col-md-6 mt-3">
    <p>This book was built by the <a class="text-light" href="https://bookdown.org">bookdown</a> R package.</p>
  </div>

</div></div>
</footer>
</body>
</html>
