<!DOCTYPE html>
<html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Chapter 13 Differential Expression | scRNAseq Analysis in R with Seurat</title>
<meta name="author" content="Monash Genomics and Bioinformatics Platform (MGBP)">
<meta name="description" content="Slides There are many different methods for calculating differential expression between groups in scRNAseq data. There are a number of review papers worth consulting on this topic. There is the...">
<meta name="generator" content="bookdown 0.41 with bs4_book()">
<meta property="og:title" content="Chapter 13 Differential Expression | scRNAseq Analysis in R with Seurat">
<meta property="og:type" content="book">
<meta property="og:description" content="Slides There are many different methods for calculating differential expression between groups in scRNAseq data. There are a number of review papers worth consulting on this topic. There is the...">
<meta name="twitter:card" content="summary">
<meta name="twitter:title" content="Chapter 13 Differential Expression | scRNAseq Analysis in R with Seurat">
<meta name="twitter:description" content="Slides There are many different methods for calculating differential expression between groups in scRNAseq data. There are a number of review papers worth consulting on this topic. There is the...">
<!-- JS --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://kit.fontawesome.com/6ecbd6c532.js" crossorigin="anonymous"></script><script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="libs/bootstrap-4.6.0/bootstrap.min.css" rel="stylesheet">
<script src="libs/bootstrap-4.6.0/bootstrap.bundle.min.js"></script><script src="libs/bs3compat-0.8.0/transition.js"></script><script src="libs/bs3compat-0.8.0/tabs.js"></script><script src="libs/bs3compat-0.8.0/bs3compat.js"></script><link href="libs/bs4_book-1.0.0/bs4_book.css" rel="stylesheet">
<script src="libs/bs4_book-1.0.0/bs4_book.js"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- CSS --><style type="text/css">
    
    div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  </style>
<link rel="stylesheet" href="style.css">
</head>
<body data-spy="scroll" data-target="#toc">

<div class="container-fluid">
<div class="row">
  <header class="col-sm-12 col-lg-3 sidebar sidebar-book"><a class="sr-only sr-only-focusable" href="#content">Skip to main content</a>

    <div class="d-flex align-items-start justify-content-between">
      <h1>
        <a href="index.html" title="">scRNAseq Analysis in R with Seurat</a>
      </h1>
      <button class="btn btn-outline-primary d-lg-none ml-2 mt-1" type="button" data-toggle="collapse" data-target="#main-nav" aria-expanded="true" aria-controls="main-nav"><i class="fas fa-bars"></i><span class="sr-only">Show table of contents</span></button>
    </div>

    <div id="main-nav" class="collapse-lg">
      <form role="search">
        <input id="search" class="form-control" type="search" placeholder="Search" aria-label="Search">
</form>

      <nav aria-label="Table of contents"><h2>Table of contents</h2>
        <ul class="book-toc list-unstyled">
<li><a class="" href="index.html"><span class="header-section-number">1</span> Getting started</a></li>
<li><a class="" href="set-up.html"><span class="header-section-number">2</span> Set up</a></li>
<li><a class="" href="schedule.html"><span class="header-section-number">3</span> Schedule</a></li>
<li><a class="" href="load.html"><span class="header-section-number">4</span> Load data</a></li>
<li class="book-part">Single Cell Analysis</li>
<li><a class="" href="qc.html"><span class="header-section-number">5</span> QC Filtering</a></li>
<li><a class="" href="norm.html"><span class="header-section-number">6</span> Normalisation</a></li>
<li><a class="" href="reducedims.html"><span class="header-section-number">7</span> PCAs and UMAPs</a></li>
<li><a class="" href="dimensionality-reduction.html"><span class="header-section-number">8</span> Dimensionality reduction</a></li>
<li><a class="" href="Harmony.html"><span class="header-section-number">9</span> Data set integration with Harmony</a></li>
<li><a class="" href="clustering.html"><span class="header-section-number">10</span> Clustering</a></li>
<li><a class="" href="clustermarkers.html"><span class="header-section-number">11</span> Cluster Markers</a></li>
<li class="book-part">Futher Analysis</li>
<li><a class="" href="singler.html"><span class="header-section-number">12</span> SingleR</a></li>
<li><a class="active" href="de2.html"><span class="header-section-number">13</span> Differential Expression</a></li>
<li><a class="" href="CellCycle.html"><span class="header-section-number">14</span> Cell cycle Assignment</a></li>
<li class="book-part">Other Resources</li>
<li><a class="" href="resources.html"><span class="header-section-number">15</span> Resources</a></li>
<li class="book-part">Seurat Object</li>
<li><a class="" href="structure.html"><span class="header-section-number">16</span> Structure</a></li>
<li><a class="" href="acknowledgements.html"><span class="header-section-number">17</span> Acknowledgements</a></li>
<li><a class="" href="session-info.html"><span class="header-section-number">18</span> Session info</a></li>
</ul>

        <div class="book-extra">
          <p><a id="book-repo" href="https://monashbioinformaticsplatform.github.io/scRNAseq_Workshop_ABACBS_2024/">View book source <i class="fab fa-gitlab"></i></a></p>
        </div>
      </nav>
</div>
  </header><main class="col-sm-12 col-md-9 col-lg-7" id="content"><div id="de2" class="section level1" number="13">
<h1>
<span class="header-section-number">13</span> Differential Expression<a class="anchor" aria-label="anchor" href="#de2"><i class="fas fa-link"></i></a>
</h1>
<p><a href="https://docs.google.com/presentation/d/1iTfkc2KgGy3d6yT9zCALZzk993pWwT9D/edit#slide=id.p1">Slides</a></p>
<p>There are many different methods for calculating differential expression between groups in scRNAseq data. There are a number of review papers worth consulting on this topic.</p>
<p>There is the <a href="https://satijalab.org/seurat/archive/v3.1/de_vignette.html">Seurat differential expression Vignette</a> which walks through the variety implemented in Seurat.</p>
<p>There is also a good discussion of useing <a href="http://bioconductor.org/books/3.15/OSCA.multisample/multi-sample-comparisons.html#creating-pseudo-bulk-samples">pseudobulk approaches</a> which is worth checking out if youre planning differential expression analyses.</p>
<div class="sourceCode" id="cb81"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">seurat_object</span><span class="op">@</span><span class="va">meta.data</span><span class="op">)</span></span>
<span><span class="co">#&gt;                     orig.ident nCount_RNA nFeature_RNA  ind</span></span>
<span><span class="co">#&gt; AGGGCGCTATTTCC-1 SeuratProject       2053          532 1256</span></span>
<span><span class="co">#&gt; GGAGACGATTCGTT-1 SeuratProject        881          392 1256</span></span>
<span><span class="co">#&gt; CACCGTTGTCGTAG-1 SeuratProject       3130         1005 1016</span></span>
<span><span class="co">#&gt; TATCGTACACGCAT-1 SeuratProject       1042          549 1488</span></span>
<span><span class="co">#&gt; TACGAGACCTATTC-1 SeuratProject       2425          777 1244</span></span>
<span><span class="co">#&gt; GTACTACTCATACG-1 SeuratProject       3951         1064 1256</span></span>
<span><span class="co">#&gt;                  stim              cell multiplets</span></span>
<span><span class="co">#&gt; AGGGCGCTATTTCC-1 stim   CD14+ Monocytes    singlet</span></span>
<span><span class="co">#&gt; GGAGACGATTCGTT-1 stim       CD4 T cells    singlet</span></span>
<span><span class="co">#&gt; CACCGTTGTCGTAG-1 ctrl FCGR3A+ Monocytes    singlet</span></span>
<span><span class="co">#&gt; TATCGTACACGCAT-1 stim           B cells    singlet</span></span>
<span><span class="co">#&gt; TACGAGACCTATTC-1 stim       CD4 T cells    singlet</span></span>
<span><span class="co">#&gt; GTACTACTCATACG-1 ctrl FCGR3A+ Monocytes    singlet</span></span>
<span><span class="co">#&gt;                  percent.mt RNA_snn_res.0.25</span></span>
<span><span class="co">#&gt; AGGGCGCTATTTCC-1  1.6336634                1</span></span>
<span><span class="co">#&gt; GGAGACGATTCGTT-1  4.8809524                0</span></span>
<span><span class="co">#&gt; CACCGTTGTCGTAG-1  1.0655473                4</span></span>
<span><span class="co">#&gt; TATCGTACACGCAT-1  3.0662710                3</span></span>
<span><span class="co">#&gt; TACGAGACCTATTC-1  1.0837849                0</span></span>
<span><span class="co">#&gt; GTACTACTCATACG-1  0.7137395                4</span></span>
<span><span class="co">#&gt;                  seurat_clusters pca_clusters</span></span>
<span><span class="co">#&gt; AGGGCGCTATTTCC-1               1            3</span></span>
<span><span class="co">#&gt; GGAGACGATTCGTT-1              14            0</span></span>
<span><span class="co">#&gt; CACCGTTGTCGTAG-1              15            7</span></span>
<span><span class="co">#&gt; TATCGTACACGCAT-1               7            5</span></span>
<span><span class="co">#&gt; TACGAGACCTATTC-1               6            0</span></span>
<span><span class="co">#&gt; GTACTACTCATACG-1              15            7</span></span>
<span><span class="co">#&gt;                  harmony_clusters RNA_snn_res.0.5</span></span>
<span><span class="co">#&gt; AGGGCGCTATTTCC-1                1               3</span></span>
<span><span class="co">#&gt; GGAGACGATTCGTT-1                0               0</span></span>
<span><span class="co">#&gt; CACCGTTGTCGTAG-1                4               9</span></span>
<span><span class="co">#&gt; TATCGTACACGCAT-1                3               6</span></span>
<span><span class="co">#&gt; TACGAGACCTATTC-1                0               0</span></span>
<span><span class="co">#&gt; GTACTACTCATACG-1                4               9</span></span>
<span><span class="co">#&gt;                  RNA_snn_res.0.1 RNA_snn_res.0.2</span></span>
<span><span class="co">#&gt; AGGGCGCTATTTCC-1               3               3</span></span>
<span><span class="co">#&gt; GGAGACGATTCGTT-1               1               0</span></span>
<span><span class="co">#&gt; CACCGTTGTCGTAG-1               2               7</span></span>
<span><span class="co">#&gt; TATCGTACACGCAT-1               4               5</span></span>
<span><span class="co">#&gt; TACGAGACCTATTC-1               1               0</span></span>
<span><span class="co">#&gt; GTACTACTCATACG-1               2               7</span></span>
<span><span class="co">#&gt;                  RNA_snn_res.0.3 RNA_snn_res.0.4</span></span>
<span><span class="co">#&gt; AGGGCGCTATTTCC-1               2               3</span></span>
<span><span class="co">#&gt; GGAGACGATTCGTT-1               0               0</span></span>
<span><span class="co">#&gt; CACCGTTGTCGTAG-1               9               9</span></span>
<span><span class="co">#&gt; TATCGTACACGCAT-1               6               6</span></span>
<span><span class="co">#&gt; TACGAGACCTATTC-1               0               0</span></span>
<span><span class="co">#&gt; GTACTACTCATACG-1               9               9</span></span>
<span><span class="co">#&gt;                  RNA_snn_res.0.6 RNA_snn_res.0.7</span></span>
<span><span class="co">#&gt; AGGGCGCTATTTCC-1               3               3</span></span>
<span><span class="co">#&gt; GGAGACGATTCGTT-1               0               0</span></span>
<span><span class="co">#&gt; CACCGTTGTCGTAG-1               9               9</span></span>
<span><span class="co">#&gt; TATCGTACACGCAT-1               6               6</span></span>
<span><span class="co">#&gt; TACGAGACCTATTC-1               0               0</span></span>
<span><span class="co">#&gt; GTACTACTCATACG-1               9               9</span></span>
<span><span class="co">#&gt;                  RNA_snn_res.0.8 RNA_snn_res.0.9</span></span>
<span><span class="co">#&gt; AGGGCGCTATTTCC-1               3               3</span></span>
<span><span class="co">#&gt; GGAGACGATTCGTT-1               0               0</span></span>
<span><span class="co">#&gt; CACCGTTGTCGTAG-1               9               9</span></span>
<span><span class="co">#&gt; TATCGTACACGCAT-1               6               6</span></span>
<span><span class="co">#&gt; TACGAGACCTATTC-1               0               0</span></span>
<span><span class="co">#&gt; GTACTACTCATACG-1               9               9</span></span>
<span><span class="co">#&gt;                  RNA_snn_res.1 RNA_snn_res.1.1</span></span>
<span><span class="co">#&gt; AGGGCGCTATTTCC-1             3               3</span></span>
<span><span class="co">#&gt; GGAGACGATTCGTT-1            10              10</span></span>
<span><span class="co">#&gt; CACCGTTGTCGTAG-1            11              11</span></span>
<span><span class="co">#&gt; TATCGTACACGCAT-1             6               6</span></span>
<span><span class="co">#&gt; TACGAGACCTATTC-1             0               0</span></span>
<span><span class="co">#&gt; GTACTACTCATACG-1            11              11</span></span>
<span><span class="co">#&gt;                  RNA_snn_res.1.2 RNA_snn_res.1.3</span></span>
<span><span class="co">#&gt; AGGGCGCTATTTCC-1               3               2</span></span>
<span><span class="co">#&gt; GGAGACGATTCGTT-1              10              12</span></span>
<span><span class="co">#&gt; CACCGTTGTCGTAG-1              11              13</span></span>
<span><span class="co">#&gt; TATCGTACACGCAT-1               6               7</span></span>
<span><span class="co">#&gt; TACGAGACCTATTC-1               0               0</span></span>
<span><span class="co">#&gt; GTACTACTCATACG-1              11              13</span></span>
<span><span class="co">#&gt;                  RNA_snn_res.1.4 RNA_snn_res.1.5</span></span>
<span><span class="co">#&gt; AGGGCGCTATTTCC-1               2               2</span></span>
<span><span class="co">#&gt; GGAGACGATTCGTT-1              12              13</span></span>
<span><span class="co">#&gt; CACCGTTGTCGTAG-1              13              14</span></span>
<span><span class="co">#&gt; TATCGTACACGCAT-1               7               8</span></span>
<span><span class="co">#&gt; TACGAGACCTATTC-1               0               0</span></span>
<span><span class="co">#&gt; GTACTACTCATACG-1              13              14</span></span>
<span><span class="co">#&gt;                  RNA_snn_res.1.6 RNA_snn_res.1.7</span></span>
<span><span class="co">#&gt; AGGGCGCTATTTCC-1               2               2</span></span>
<span><span class="co">#&gt; GGAGACGATTCGTT-1              13              13</span></span>
<span><span class="co">#&gt; CACCGTTGTCGTAG-1              14              14</span></span>
<span><span class="co">#&gt; TATCGTACACGCAT-1               7               8</span></span>
<span><span class="co">#&gt; TACGAGACCTATTC-1               0               0</span></span>
<span><span class="co">#&gt; GTACTACTCATACG-1              14              14</span></span>
<span><span class="co">#&gt;                  RNA_snn_res.1.8 RNA_snn_res.1.9</span></span>
<span><span class="co">#&gt; AGGGCGCTATTTCC-1               1               2</span></span>
<span><span class="co">#&gt; GGAGACGATTCGTT-1              14              14</span></span>
<span><span class="co">#&gt; CACCGTTGTCGTAG-1              15              15</span></span>
<span><span class="co">#&gt; TATCGTACACGCAT-1               9              10</span></span>
<span><span class="co">#&gt; TACGAGACCTATTC-1               7               7</span></span>
<span><span class="co">#&gt; GTACTACTCATACG-1              15              15</span></span>
<span><span class="co">#&gt;                  RNA_snn_res.2 Naive_CD4_T1  cell_label</span></span>
<span><span class="co">#&gt; AGGGCGCTATTTCC-1             1   -0.3540023        &lt;NA&gt;</span></span>
<span><span class="co">#&gt; GGAGACGATTCGTT-1            14    2.1376158 Naive_CD4_T</span></span>
<span><span class="co">#&gt; CACCGTTGTCGTAG-1            15   -1.1487836        &lt;NA&gt;</span></span>
<span><span class="co">#&gt; TATCGTACACGCAT-1             7    0.5557941        &lt;NA&gt;</span></span>
<span><span class="co">#&gt; TACGAGACCTATTC-1             6    1.4250065 Naive_CD4_T</span></span>
<span><span class="co">#&gt; GTACTACTCATACG-1            15   -0.2082793        &lt;NA&gt;</span></span>
<span><span class="co">#&gt;                  SingleR.labels</span></span>
<span><span class="co">#&gt; AGGGCGCTATTTCC-1       Monocyte</span></span>
<span><span class="co">#&gt; GGAGACGATTCGTT-1        T_cells</span></span>
<span><span class="co">#&gt; CACCGTTGTCGTAG-1       Monocyte</span></span>
<span><span class="co">#&gt; TATCGTACACGCAT-1    Neutrophils</span></span>
<span><span class="co">#&gt; TACGAGACCTATTC-1        T_cells</span></span>
<span><span class="co">#&gt; GTACTACTCATACG-1       Monocyte</span></span></code></pre></div>
<p>How cells from each condition do we have?</p>
<div class="sourceCode" id="cb82"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/table.html">table</a></span><span class="op">(</span><span class="va">seurat_object</span><span class="op">$</span><span class="va">stim</span><span class="op">)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; ctrl stim </span></span>
<span><span class="co">#&gt; 2378 2499</span></span></code></pre></div>
<p>How many cells per individuals per group?</p>
<div class="sourceCode" id="cb83"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/table.html">table</a></span><span class="op">(</span><span class="va">seurat_object</span><span class="op">$</span><span class="va">ind</span>, <span class="va">seurat_object</span><span class="op">$</span><span class="va">stim</span><span class="op">)</span></span>
<span><span class="co">#&gt;       </span></span>
<span><span class="co">#&gt;        ctrl stim</span></span>
<span><span class="co">#&gt;   101   175  226</span></span>
<span><span class="co">#&gt;   107   115  106</span></span>
<span><span class="co">#&gt;   1015  503  489</span></span>
<span><span class="co">#&gt;   1016  335  348</span></span>
<span><span class="co">#&gt;   1039   87  100</span></span>
<span><span class="co">#&gt;   1244  373  311</span></span>
<span><span class="co">#&gt;   1256  384  385</span></span>
<span><span class="co">#&gt;   1488  406  534</span></span></code></pre></div>
<p>And for each sample, how many of each cell type has been classified?</p>
<div class="sourceCode" id="cb84"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/table.html">table</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste</a></span><span class="op">(</span><span class="va">seurat_object</span><span class="op">$</span><span class="va">ind</span>,<span class="va">seurat_object</span><span class="op">$</span><span class="va">stim</span><span class="op">)</span>, <span class="va">seurat_object</span><span class="op">$</span><span class="va">cell</span><span class="op">)</span></span>
<span><span class="co">#&gt;            </span></span>
<span><span class="co">#&gt;             B cells CD14+ Monocytes CD4 T cells CD8 T cells</span></span>
<span><span class="co">#&gt;   101 ctrl       23              47          60          15</span></span>
<span><span class="co">#&gt;   101 stim       30              52          83          17</span></span>
<span><span class="co">#&gt;   1015 ctrl      79             149         142          44</span></span>
<span><span class="co">#&gt;   1015 stim      68             149         148          21</span></span>
<span><span class="co">#&gt;   1016 ctrl      21              71          82         107</span></span>
<span><span class="co">#&gt;   1016 stim      29              65          65         109</span></span>
<span><span class="co">#&gt;   1039 ctrl       7              32          36           6</span></span>
<span><span class="co">#&gt;   1039 stim       6              26          48           5</span></span>
<span><span class="co">#&gt;   107 ctrl        9              50          32           6</span></span>
<span><span class="co">#&gt;   107 stim        9              35          43           1</span></span>
<span><span class="co">#&gt;   1244 ctrl      23              85         202           8</span></span>
<span><span class="co">#&gt;   1244 stim      18              58         191           4</span></span>
<span><span class="co">#&gt;   1256 ctrl      32              80         176          26</span></span>
<span><span class="co">#&gt;   1256 stim      42              70         194          25</span></span>
<span><span class="co">#&gt;   1488 ctrl      36              59         246          13</span></span>
<span><span class="co">#&gt;   1488 stim      59              59         319          15</span></span>
<span><span class="co">#&gt;            </span></span>
<span><span class="co">#&gt;             Dendritic cells FCGR3A+ Monocytes</span></span>
<span><span class="co">#&gt;   101 ctrl                4                11</span></span>
<span><span class="co">#&gt;   101 stim                6                23</span></span>
<span><span class="co">#&gt;   1015 ctrl               3                49</span></span>
<span><span class="co">#&gt;   1015 stim              17                43</span></span>
<span><span class="co">#&gt;   1016 ctrl               4                21</span></span>
<span><span class="co">#&gt;   1016 stim               2                32</span></span>
<span><span class="co">#&gt;   1039 ctrl               1                 3</span></span>
<span><span class="co">#&gt;   1039 stim               1                 6</span></span>
<span><span class="co">#&gt;   107 ctrl                3                12</span></span>
<span><span class="co">#&gt;   107 stim                2                 5</span></span>
<span><span class="co">#&gt;   1244 ctrl               8                19</span></span>
<span><span class="co">#&gt;   1244 stim               6                 4</span></span>
<span><span class="co">#&gt;   1256 ctrl               6                20</span></span>
<span><span class="co">#&gt;   1256 stim               3                11</span></span>
<span><span class="co">#&gt;   1488 ctrl               8                25</span></span>
<span><span class="co">#&gt;   1488 stim              12                28</span></span>
<span><span class="co">#&gt;            </span></span>
<span><span class="co">#&gt;             Megakaryocytes NK cells</span></span>
<span><span class="co">#&gt;   101 ctrl               4       11</span></span>
<span><span class="co">#&gt;   101 stim               1       14</span></span>
<span><span class="co">#&gt;   1015 ctrl              5       32</span></span>
<span><span class="co">#&gt;   1015 stim              5       38</span></span>
<span><span class="co">#&gt;   1016 ctrl              3       26</span></span>
<span><span class="co">#&gt;   1016 stim              1       45</span></span>
<span><span class="co">#&gt;   1039 ctrl              1        1</span></span>
<span><span class="co">#&gt;   1039 stim              3        5</span></span>
<span><span class="co">#&gt;   107 ctrl               0        3</span></span>
<span><span class="co">#&gt;   107 stim               0       11</span></span>
<span><span class="co">#&gt;   1244 ctrl              3       25</span></span>
<span><span class="co">#&gt;   1244 stim              4       26</span></span>
<span><span class="co">#&gt;   1256 ctrl              1       43</span></span>
<span><span class="co">#&gt;   1256 stim              7       33</span></span>
<span><span class="co">#&gt;   1488 ctrl              4       15</span></span>
<span><span class="co">#&gt;   1488 stim              6       36</span></span></code></pre></div>
<div id="prefiltering" class="section level2" number="13.1">
<h2>
<span class="header-section-number">13.1</span> Prefiltering<a class="anchor" aria-label="anchor" href="#prefiltering"><i class="fas fa-link"></i></a>
</h2>
<div id="why-do-we-need-to-do-this-8" class="section level4 unnumbered rational">
<h4>Why do we need to do this?<a class="anchor" aria-label="anchor" href="#why-do-we-need-to-do-this-8"><i class="fas fa-link"></i></a>
</h4>
<p>If expression is below a certain level, it will be almost impossible to see any differential expression.</p>
</div>
<div id="section-14" class="section level4 unnumbered">
<h4 class="unnumbered"><a class="anchor" aria-label="anchor" href="#section-14"><i class="fas fa-link"></i></a></h4>
<p>When doing differential expression, you generally ignore genes with low expression.
In single cell datasets, there are many genes like this. Filtering here to make our dataset smaller so it runs quicker, and there is less aggressive correction for multiple hypotheses.</p>
<p>How many genes before filtering?</p>
<div class="sourceCode" id="cb85"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">seurat_object</span></span>
<span><span class="co">#&gt; An object of class Seurat </span></span>
<span><span class="co">#&gt; 35635 features across 4877 samples within 1 assay </span></span>
<span><span class="co">#&gt; Active assay: RNA (35635 features, 2000 variable features)</span></span>
<span><span class="co">#&gt;  3 layers present: counts, data, scale.data</span></span>
<span><span class="co">#&gt;  4 dimensional reductions calculated: pca, umap, harmony, umap_harmony</span></span></code></pre></div>
<p>How many copies of each gene are there?</p>
<div class="sourceCode" id="cb86"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">total_per_gene</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colSums.html">rowSums</a></span><span class="op">(</span><span class="fu"><a href="https://mojaveazure.github.io/seurat-object/reference/AssayData.html">GetAssayData</a></span><span class="op">(</span><span class="va">seurat_object</span>, assay<span class="op">=</span><span class="st">'RNA'</span>, slot<span class="op">=</span><span class="st">'counts'</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; Warning: The `slot` argument of `GetAssayData()` is deprecated as of SeuratObject 5.0.0.</span></span>
<span><span class="co">#&gt; ℹ Please use the `layer` argument instead.</span></span>
<span><span class="co">#&gt; This warning is displayed once every 8 hours.</span></span>
<span><span class="co">#&gt; Call `lifecycle::last_lifecycle_warnings()` to see where this warning was generated.</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/hist.html">hist</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html">log10</a></span><span class="op">(</span><span class="va">total_per_gene</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<div class="inline-figure"><img src="scRNAseqInR_ABACBS_2024_Doco_files/figure-html/unnamed-chunk-48-1.png" width="672"></div>
<p>Lets keep only those genes with at least 50 copies across the entire experiment.</p>
<div class="sourceCode" id="cb87"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">seurat_object</span><span class="op">&lt;-</span> <span class="va">seurat_object</span><span class="op">[</span><span class="va">total_per_gene</span> <span class="op">&gt;=</span> <span class="fl">50</span>, <span class="op">]</span> </span></code></pre></div>
<p>How many genes after filtering?</p>
<div class="sourceCode" id="cb88"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">seurat_object</span></span>
<span><span class="co">#&gt; An object of class Seurat </span></span>
<span><span class="co">#&gt; 7200 features across 4877 samples within 1 assay </span></span>
<span><span class="co">#&gt; Active assay: RNA (7200 features, 1236 variable features)</span></span>
<span><span class="co">#&gt;  3 layers present: counts, data, scale.data</span></span>
<span><span class="co">#&gt;  4 dimensional reductions calculated: pca, umap, harmony, umap_harmony</span></span></code></pre></div>
<p>We might like to see the effect of IFN-beta stimulation on each cell type individually. For the purposes of this workshop, just going to test one cell type; CD14+ Monocytes</p>
<p>An easy way is to subset the object.</p>
<div class="sourceCode" id="cb89"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Set idents to 'cell' column.</span></span>
<span><span class="fu"><a href="https://mojaveazure.github.io/seurat-object/reference/Idents.html">Idents</a></span><span class="op">(</span><span class="va">seurat_object</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">seurat_object</span><span class="op">$</span><span class="va">cell</span></span>
<span><span class="fu"><a href="https://satijalab.org/seurat/reference/DimPlot.html">DimPlot</a></span><span class="op">(</span><span class="va">seurat_object</span>, reduction <span class="op">=</span> <span class="st">"umap_harmony"</span><span class="op">)</span></span></code></pre></div>
<div class="inline-figure"><img src="scRNAseqInR_ABACBS_2024_Doco_files/figure-html/unnamed-chunk-51-1.png" width="672"></div>
<div class="sourceCode" id="cb90"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">seurat_object_celltype</span> <span class="op">&lt;-</span> <span class="va">seurat_object</span><span class="op">[</span>, <span class="va">seurat_object</span><span class="op">$</span><span class="va">cell</span> <span class="op">==</span> <span class="st">"CD14+ Monocytes"</span> <span class="op">]</span></span>
<span><span class="fu"><a href="https://satijalab.org/seurat/reference/DimPlot.html">DimPlot</a></span><span class="op">(</span><span class="va">seurat_object_celltype</span>, reduction <span class="op">=</span> <span class="st">"umap_harmony"</span><span class="op">)</span></span></code></pre></div>
<div class="inline-figure"><img src="scRNAseqInR_ABACBS_2024_Doco_files/figure-html/unnamed-chunk-51-2.png" width="672"></div>
</div>
</div>
<div id="default-wilcox-test" class="section level2" number="13.2">
<h2>
<span class="header-section-number">13.2</span> Default Wilcox test<a class="anchor" aria-label="anchor" href="#default-wilcox-test"><i class="fas fa-link"></i></a>
</h2>
<p>To run this test, we change the Idents to the factor(column) we want to test. In this case, that’s ‘stim’.</p>
<div class="sourceCode" id="cb91"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Change Ident to Condition</span></span>
<span><span class="fu"><a href="https://mojaveazure.github.io/seurat-object/reference/Idents.html">Idents</a></span><span class="op">(</span><span class="va">seurat_object_celltype</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">seurat_object_celltype</span><span class="op">$</span><span class="va">stim</span></span>
<span></span>
<span><span class="co"># default, wilcox test</span></span>
<span><span class="va">de_result_wilcox</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/FindMarkers.html">FindMarkers</a></span><span class="op">(</span><span class="va">seurat_object_celltype</span>, </span>
<span>            ident.1 <span class="op">=</span> <span class="st">'stim'</span>,</span>
<span>            ident.2 <span class="op">=</span> <span class="st">'ctrl'</span>,</span>
<span>            logfc.threshold <span class="op">=</span> <span class="fl">0</span>, <span class="co"># Give me ALL results</span></span>
<span>            min.pct <span class="op">=</span> <span class="fl">0</span></span>
<span>            <span class="op">)</span></span>
<span></span>
<span><span class="co"># Add average expression for plotting</span></span>
<span><span class="va">de_result_wilcox</span><span class="op">$</span><span class="va">AveExpr</span><span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colSums.html">rowMeans</a></span><span class="op">(</span><span class="va">seurat_object_celltype</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">de_result_wilcox</span><span class="op">)</span>,<span class="op">]</span><span class="op">)</span></span></code></pre></div>
<p>Look at the top differentially expressed genes.</p>
<div class="sourceCode" id="cb92"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">de_result_wilcox</span><span class="op">)</span></span>
<span><span class="co">#&gt;                 p_val avg_log2FC pct.1 pct.2     p_val_adj</span></span>
<span><span class="co">#&gt; RSAD2   8.471603e-196   6.797508 0.979 0.044 6.099554e-192</span></span>
<span><span class="co">#&gt; TNFSF10 3.399691e-195   6.559853 0.982 0.056 2.447777e-191</span></span>
<span><span class="co">#&gt; CXCL10  5.809758e-195   8.034949 0.975 0.038 4.183025e-191</span></span>
<span><span class="co">#&gt; IFIT3   7.214495e-195   6.234703 0.982 0.051 5.194437e-191</span></span>
<span><span class="co">#&gt; IFIT1   1.224398e-188   6.764699 0.953 0.026 8.815662e-185</span></span>
<span><span class="co">#&gt; ISG15   1.228738e-186   6.377243 1.000 0.323 8.846914e-183</span></span>
<span><span class="co">#&gt;            AveExpr</span></span>
<span><span class="co">#&gt; RSAD2   0.04196286</span></span>
<span><span class="co">#&gt; TNFSF10 0.51655832</span></span>
<span><span class="co">#&gt; CXCL10  3.25184139</span></span>
<span><span class="co">#&gt; IFIT3   0.02032398</span></span>
<span><span class="co">#&gt; IFIT1   0.01254584</span></span>
<span><span class="co">#&gt; ISG15   0.14165156</span></span></code></pre></div>
<div class="sourceCode" id="cb93"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">p1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="va">de_result_wilcox</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x<span class="op">=</span><span class="va">AveExpr</span>, y<span class="op">=</span><span class="va">avg_log2FC</span>, col<span class="op">=</span><span class="va">p_val_adj</span> <span class="op">&lt;</span> <span class="fl">0.05</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html">scale_colour_manual</a></span><span class="op">(</span>values<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">'TRUE'</span><span class="op">=</span><span class="st">"red"</span>,<span class="st">'FALSE'</span><span class="op">=</span><span class="st">"black"</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_bw</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">ggtitle</a></span><span class="op">(</span><span class="st">"Wilcox Test"</span><span class="op">)</span></span>
<span></span>
<span></span>
<span><span class="va">p2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="va">de_result_wilcox</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x<span class="op">=</span><span class="va">avg_log2FC</span>, y<span class="op">=</span><span class="op">-</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html">log10</a></span><span class="op">(</span><span class="va">p_val</span><span class="op">)</span>, col<span class="op">=</span><span class="va">p_val_adj</span> <span class="op">&lt;</span> <span class="fl">0.05</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html">scale_colour_manual</a></span><span class="op">(</span>values<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">'TRUE'</span><span class="op">=</span><span class="st">"red"</span>,<span class="st">'FALSE'</span><span class="op">=</span><span class="st">"black"</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_bw</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">ggtitle</a></span><span class="op">(</span><span class="st">"Wilcox Test (Volcano)"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">p1</span> <span class="op">+</span> <span class="va">p2</span></span></code></pre></div>
<div class="inline-figure"><img src="scRNAseqInR_ABACBS_2024_Doco_files/figure-html/unnamed-chunk-54-1.png" width="672"></div>
</div>
<div id="seurat-negative-binomial" class="section level2" number="13.3">
<h2>
<span class="header-section-number">13.3</span> Seurat Negative binomial<a class="anchor" aria-label="anchor" href="#seurat-negative-binomial"><i class="fas fa-link"></i></a>
</h2>
<p>Negative binonial test is run almost the same way - just need to specify it under ‘test.use’</p>
<div class="sourceCode" id="cb94"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="co"># Change Ident to Condition</span></span>
<span><span class="fu"><a href="https://mojaveazure.github.io/seurat-object/reference/Idents.html">Idents</a></span><span class="op">(</span><span class="va">seurat_object_celltype</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">seurat_object_celltype</span><span class="op">$</span><span class="va">stim</span></span>
<span></span>
<span><span class="co"># default, wilcox test</span></span>
<span><span class="va">de_result_negbinom</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/FindMarkers.html">FindMarkers</a></span><span class="op">(</span><span class="va">seurat_object_celltype</span>, </span>
<span>            test.use<span class="op">=</span><span class="st">"negbinom"</span>, <span class="co"># Choose a different test.</span></span>
<span>            ident.1 <span class="op">=</span> <span class="st">'stim'</span>,</span>
<span>            ident.2 <span class="op">=</span> <span class="st">'ctrl'</span>,</span>
<span>            logfc.threshold <span class="op">=</span> <span class="fl">0</span>, <span class="co"># Give me ALL results</span></span>
<span>            min.pct <span class="op">=</span> <span class="fl">0</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Add average expression for plotting</span></span>
<span><span class="va">de_result_negbinom</span><span class="op">$</span><span class="va">AveExpr</span><span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colSums.html">rowMeans</a></span><span class="op">(</span><span class="va">seurat_object_celltype</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">de_result_negbinom</span><span class="op">)</span>,<span class="op">]</span><span class="op">)</span></span></code></pre></div>
<p>Look at the top differentially expressed genes.</p>
<div class="sourceCode" id="cb95"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">de_result_negbinom</span><span class="op">)</span></span>
<span><span class="co">#&gt;          p_val avg_log2FC pct.1 pct.2 p_val_adj    AveExpr</span></span>
<span><span class="co">#&gt; CXCL10       0   8.034949 0.975 0.038         0 0.04196286</span></span>
<span><span class="co">#&gt; CCL8         0   8.160755 0.912 0.023         0 0.51655832</span></span>
<span><span class="co">#&gt; LY6E         0   4.880093 0.979 0.134         0 3.25184139</span></span>
<span><span class="co">#&gt; APOBEC3A     0   4.745713 0.996 0.262         0 0.02032398</span></span>
<span><span class="co">#&gt; IFITM3       0   4.675898 0.998 0.271         0 0.01254584</span></span>
<span><span class="co">#&gt; IFI6         0   3.935122 0.982 0.257         0 0.14165156</span></span></code></pre></div>
<div class="sourceCode" id="cb96"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">p1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="va">de_result_negbinom</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x<span class="op">=</span><span class="va">AveExpr</span>, y<span class="op">=</span><span class="va">avg_log2FC</span>, col<span class="op">=</span><span class="va">p_val_adj</span> <span class="op">&lt;</span> <span class="fl">0.05</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html">scale_colour_manual</a></span><span class="op">(</span>values<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">'TRUE'</span><span class="op">=</span><span class="st">"red"</span>,<span class="st">'FALSE'</span><span class="op">=</span><span class="st">"black"</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_bw</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">ggtitle</a></span><span class="op">(</span><span class="st">"Negative Bionomial Test"</span><span class="op">)</span></span>
<span></span>
<span></span>
<span><span class="va">p2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="va">de_result_negbinom</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x<span class="op">=</span><span class="va">avg_log2FC</span>, y<span class="op">=</span><span class="op">-</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html">log10</a></span><span class="op">(</span><span class="va">p_val</span><span class="op">)</span>, col<span class="op">=</span><span class="va">p_val_adj</span> <span class="op">&lt;</span> <span class="fl">0.05</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html">scale_colour_manual</a></span><span class="op">(</span>values<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">'TRUE'</span><span class="op">=</span><span class="st">"red"</span>,<span class="st">'FALSE'</span><span class="op">=</span><span class="st">"black"</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_bw</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">ggtitle</a></span><span class="op">(</span><span class="st">"Negative Bionomial Test (Volcano)"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">p1</span> <span class="op">+</span> <span class="va">p2</span></span></code></pre></div>
<div class="inline-figure"><img src="scRNAseqInR_ABACBS_2024_Doco_files/figure-html/unnamed-chunk-57-1.png" width="672"></div>
</div>
<div id="pseudobulk" class="section level2" number="13.4">
<h2>
<span class="header-section-number">13.4</span> Pseudobulk<a class="anchor" aria-label="anchor" href="#pseudobulk"><i class="fas fa-link"></i></a>
</h2>
<p>Pseudobulk analysis is an option where you have biological replicates. It is essentially pooling the individual cell counts and treating your expreiment like a bulk RNAseq.</p>
<p>First, you need to build a pseudobulk matrix - the <code><a href="https://satijalab.org/seurat/reference/AggregateExpression.html">AggregateExpression()</a></code> function can do this, once you set the ‘Idents’ of your seurat object to your grouping factor (here, thats a combination of individual+treatment called ‘sample’, instead of the ‘stim’ treatment column).</p>
<div class="sourceCode" id="cb97"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Tools for bulk differential expression</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://bioinf.wehi.edu.au/limma/">limma</a></span><span class="op">)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Attaching package: 'limma'</span></span>
<span><span class="co">#&gt; The following object is masked from 'package:BiocGenerics':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     plotMA</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://bioinf.wehi.edu.au/edgeR/">edgeR</a></span><span class="op">)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Attaching package: 'edgeR'</span></span>
<span><span class="co">#&gt; The following object is masked from 'package:SingleCellExperiment':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     cpm</span></span>
<span></span>
<span></span>
<span><span class="co"># Change idents to ind for grouping.</span></span>
<span><span class="va">seurat_object_celltype</span><span class="op">$</span><span class="va">sample</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html">factor</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste</a></span><span class="op">(</span><span class="va">seurat_object_celltype</span><span class="op">$</span><span class="va">stim</span>, <span class="va">seurat_object_celltype</span><span class="op">$</span><span class="va">ind</span>, sep<span class="op">=</span><span class="st">"_"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://mojaveazure.github.io/seurat-object/reference/Idents.html">Idents</a></span><span class="op">(</span><span class="va">seurat_object_celltype</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">seurat_object_celltype</span><span class="op">$</span><span class="va">sample</span></span>
<span></span>
<span><span class="co"># THen pool together counts in those groups</span></span>
<span><span class="co"># AggregateExperssion returns a list of matricies - one for each assay requested (even just requesting one)</span></span>
<span><span class="va">pseudobulk_matrix_list</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/AggregateExpression.html">AggregateExpression</a></span><span class="op">(</span> <span class="va">seurat_object_celltype</span>,  slot <span class="op">=</span> <span class="st">'counts'</span>, assays<span class="op">=</span><span class="st">'RNA'</span> <span class="op">)</span></span>
<span><span class="co">#&gt; Names of identity class contain underscores ('_'), replacing with dashes ('-')</span></span>
<span><span class="co">#&gt; This message is displayed once every 8 hours.</span></span>
<span><span class="va">pseudobulk_matrix</span>      <span class="op">&lt;-</span> <span class="va">pseudobulk_matrix_list</span><span class="op">[[</span><span class="st">'RNA'</span><span class="op">]</span><span class="op">]</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">pseudobulk_matrix</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/character.html">as.character</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">pseudobulk_matrix</span><span class="op">)</span><span class="op">)</span> <span class="co"># Changes colnames to simple text</span></span>
<span><span class="va">pseudobulk_matrix</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span>,<span class="fl">1</span><span class="op">:</span><span class="fl">4</span><span class="op">]</span></span>
<span><span class="co">#&gt; 5 x 4 sparse Matrix of class "dgCMatrix"</span></span>
<span><span class="co">#&gt;          ctrl-101 ctrl-1015 ctrl-1016 ctrl-1039</span></span>
<span><span class="co">#&gt; NOC2L           2         7         .         .</span></span>
<span><span class="co">#&gt; HES4            .         3         2         1</span></span>
<span><span class="co">#&gt; ISG15          31       185       234        41</span></span>
<span><span class="co">#&gt; TNFRSF18        .         3         4         2</span></span>
<span><span class="co">#&gt; TNFRSF4         .         2         .         .</span></span></code></pre></div>
<p>Now it looks like a bulk RNAseq experiment, so treat it like one.</p>
<p>We can use the popular <code>limma</code> package for differential expression. Here is one <a href="https://ucdavis-bioinformatics-training.github.io/2018-June-RNA-Seq-Workshop/thursday/DE.html">tutorial</a>, and the hefty reference manual is hosted by <a href="https://bioconductor.org/packages/release/bioc/html/limma.html">bioconductor</a>.</p>
<p>In brief, this code below constructs a linear model for this experiment that accounts for the variation in individuals and treatment. It then tests for differential expression between ‘stim’ and ‘ctrl’ groups.</p>
<div class="sourceCode" id="cb98"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">dge</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/edgeR/man/DGEList.html">DGEList</a></span><span class="op">(</span><span class="va">pseudobulk_matrix</span><span class="op">)</span></span>
<span><span class="va">dge</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/edgeR/man/calcNormFactors.html">calcNormFactors</a></span><span class="op">(</span><span class="va">dge</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Remove _ or - and everything after it - yeilds stim group</span></span>
<span><span class="va">stim</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html">gsub</a></span><span class="op">(</span><span class="st">"[-_].*"</span>,<span class="st">""</span>,<span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">pseudobulk_matrix</span><span class="op">)</span><span class="op">)</span> </span>
<span></span>
<span><span class="co"># Removing everything before the _ or - for the individual, then converting those numerical ind explictiy to text. Else limma will treat them as numbers!</span></span>
<span><span class="va">ind</span>  <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/character.html">as.character</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/grep.html">gsub</a></span><span class="op">(</span><span class="st">".*[-_]"</span>,<span class="st">""</span>,<span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">pseudobulk_matrix</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> </span>
<span></span>
<span><span class="va">design</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/model.matrix.html">model.matrix</a></span><span class="op">(</span> <span class="op">~</span><span class="fl">0</span> <span class="op">+</span> <span class="va">stim</span> <span class="op">+</span> <span class="va">ind</span><span class="op">)</span></span>
<span><span class="va">vm</span>  <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/limma/man/voom.html">voom</a></span><span class="op">(</span><span class="va">dge</span>, design <span class="op">=</span> <span class="va">design</span>, plot <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="va">fit</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/limma/man/lmFit.html">lmFit</a></span><span class="op">(</span><span class="va">vm</span>, design <span class="op">=</span> <span class="va">design</span><span class="op">)</span></span>
<span></span>
<span><span class="va">contrasts</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/limma/man/makeContrasts.html">makeContrasts</a></span><span class="op">(</span><span class="va">stimstim</span> <span class="op">-</span> <span class="va">stimctrl</span>, levels<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/stats/coef.html">coef</a></span><span class="op">(</span><span class="va">fit</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">fit</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/limma/man/contrasts.fit.html">contrasts.fit</a></span><span class="op">(</span><span class="va">fit</span>, <span class="va">contrasts</span><span class="op">)</span></span>
<span><span class="va">fit</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/limma/man/ebayes.html">eBayes</a></span><span class="op">(</span><span class="va">fit</span><span class="op">)</span></span>
<span></span>
<span><span class="va">de_result_pseudobulk</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/limma/man/toptable.html">topTable</a></span><span class="op">(</span><span class="va">fit</span>, n <span class="op">=</span> <span class="cn">Inf</span>, adjust.method <span class="op">=</span> <span class="st">"BH"</span><span class="op">)</span></span>
<span><span class="va">de_result_pseudobulk</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/arrange.html">arrange</a></span><span class="op">(</span><span class="va">de_result_pseudobulk</span> , <span class="va">adj.P.Val</span><span class="op">)</span></span></code></pre></div>
<p>Look at the significantly differentially expressed genes:</p>
<div class="sourceCode" id="cb99"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">de_result_pseudobulk</span><span class="op">)</span></span>
<span><span class="co">#&gt;           logFC   AveExpr        t      P.Value</span></span>
<span><span class="co">#&gt; ISG20  5.161336 10.313033 35.72209 9.838370e-31</span></span>
<span><span class="co">#&gt; ISG15  6.932960 11.897783 34.63778 3.049704e-30</span></span>
<span><span class="co">#&gt; CXCL11 9.050139  7.263356 32.46019 3.282727e-29</span></span>
<span><span class="co">#&gt; IFIT3  6.987552  8.423467 28.97126 2.052443e-27</span></span>
<span><span class="co">#&gt; HERC5  6.999781  5.604886 28.65574 3.050588e-27</span></span>
<span><span class="co">#&gt; CXCL10 8.708460  9.715253 28.25406 5.081902e-27</span></span>
<span><span class="co">#&gt;           adj.P.Val        B</span></span>
<span><span class="co">#&gt; ISG20  7.083626e-27 60.13277</span></span>
<span><span class="co">#&gt; ISG15  1.097894e-26 59.02083</span></span>
<span><span class="co">#&gt; CXCL11 7.878545e-26 55.45779</span></span>
<span><span class="co">#&gt; IFIT3  3.694397e-24 52.01990</span></span>
<span><span class="co">#&gt; HERC5  4.392847e-24 51.18902</span></span>
<span><span class="co">#&gt; CXCL10 5.629458e-24 51.32019</span></span></code></pre></div>
<div class="sourceCode" id="cb100"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">p1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="va">de_result_pseudobulk</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x<span class="op">=</span><span class="va">AveExpr</span>, y<span class="op">=</span><span class="va">logFC</span>, col<span class="op">=</span><span class="va">adj.P.Val</span> <span class="op">&lt;</span> <span class="fl">0.05</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html">scale_colour_manual</a></span><span class="op">(</span>values<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">'TRUE'</span><span class="op">=</span><span class="st">"red"</span>,<span class="st">'FALSE'</span><span class="op">=</span><span class="st">"black"</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_bw</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">ggtitle</a></span><span class="op">(</span><span class="st">"Pseudobulk"</span><span class="op">)</span></span>
<span></span>
<span></span>
<span><span class="va">p2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="va">de_result_pseudobulk</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x<span class="op">=</span><span class="va">logFC</span>, y<span class="op">=</span><span class="op">-</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html">log10</a></span><span class="op">(</span><span class="va">P.Value</span><span class="op">)</span>, col<span class="op">=</span><span class="va">adj.P.Val</span> <span class="op">&lt;</span> <span class="fl">0.05</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html">scale_colour_manual</a></span><span class="op">(</span>values<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">'TRUE'</span><span class="op">=</span><span class="st">"red"</span>,<span class="st">'FALSE'</span><span class="op">=</span><span class="st">"black"</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_bw</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">ggtitle</a></span><span class="op">(</span><span class="st">"Pseudobulk Test (Volcano)"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">p1</span> <span class="op">+</span> <span class="va">p2</span></span></code></pre></div>
<div class="inline-figure"><img src="scRNAseqInR_ABACBS_2024_Doco_files/figure-html/unnamed-chunk-61-1.png" width="672"></div>
<div id="discussion" class="section level4 unnumbered challenge">
<h4>Discussion<a class="anchor" aria-label="anchor" href="#discussion"><i class="fas fa-link"></i></a>
</h4>
<p>These methods give different results. How would you decide which to use? How could you check an individual gene?</p>
</div>
<div id="section-15" class="section level4 unnumbered">
<h4 class="unnumbered"><a class="anchor" aria-label="anchor" href="#section-15"><i class="fas fa-link"></i></a></h4>

</div>
</div>
</div>
  <div class="chapter-nav">
<div class="prev"><a href="singler.html"><span class="header-section-number">12</span> SingleR</a></div>
<div class="next"><a href="CellCycle.html"><span class="header-section-number">14</span> Cell cycle Assignment</a></div>
</div></main><div class="col-md-3 col-lg-2 d-none d-md-block sidebar sidebar-chapter">
    <nav id="toc" data-toggle="toc" aria-label="On this page"><h2>On this page</h2>
      <ul class="nav navbar-nav">
<li><a class="nav-link" href="#de2"><span class="header-section-number">13</span> Differential Expression</a></li>
<li><a class="nav-link" href="#prefiltering"><span class="header-section-number">13.1</span> Prefiltering</a></li>
<li><a class="nav-link" href="#default-wilcox-test"><span class="header-section-number">13.2</span> Default Wilcox test</a></li>
<li><a class="nav-link" href="#seurat-negative-binomial"><span class="header-section-number">13.3</span> Seurat Negative binomial</a></li>
<li><a class="nav-link" href="#pseudobulk"><span class="header-section-number">13.4</span> Pseudobulk</a></li>
</ul>

      <div class="book-extra">
        <ul class="list-unstyled">
<li><a id="book-source" href="https://monashbioinformaticsplatform.github.io/scRNAseq_Workshop_ABACBS_2024//blob/master/04-10-de2.Rmd">View source <i class="fab fa-gitlab"></i></a></li>
          <li><a id="book-edit" href="https://monashbioinformaticsplatform.github.io/scRNAseq_Workshop_ABACBS_2024//edit/master/04-10-de2.Rmd">Edit this page <i class="fab fa-gitlab"></i></a></li>
        </ul>
</div>
    </nav>
</div>

</div>
</div> <!-- .container -->

<footer class="bg-primary text-light mt-5"><div class="container"><div class="row">

  <div class="col-12 col-md-6 mt-3">
    <p>"<strong>scRNAseq Analysis in R with Seurat</strong>" was written by Monash Genomics and Bioinformatics Platform (MGBP). It was last built on Compiled: November 07, 2024.</p>
  </div>

  <div class="col-12 col-md-6 mt-3">
    <p>This book was built by the <a class="text-light" href="https://bookdown.org">bookdown</a> R package.</p>
  </div>

</div></div>
</footer>
</body>
</html>
