<!DOCTYPE html>
<html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Chapter 3 M1 Installation | scRNAseq Analysis in R with Seurat</title>
<meta name="description" content="NOTE: THESE INSTRUCTIONS ARE ONLY FOR M1 MACS, please use the installation instructions if you do not have an M1 Mac. We strongly recommend that you use one of the provided Virtual Machines (VM)...">
<meta name="generator" content="bookdown 0.36 with bs4_book()">
<meta property="og:title" content="Chapter 3 M1 Installation | scRNAseq Analysis in R with Seurat">
<meta property="og:type" content="book">
<meta property="og:description" content="NOTE: THESE INSTRUCTIONS ARE ONLY FOR M1 MACS, please use the installation instructions if you do not have an M1 Mac. We strongly recommend that you use one of the provided Virtual Machines (VM)...">
<meta name="twitter:card" content="summary">
<meta name="twitter:title" content="Chapter 3 M1 Installation | scRNAseq Analysis in R with Seurat">
<meta name="twitter:description" content="NOTE: THESE INSTRUCTIONS ARE ONLY FOR M1 MACS, please use the installation instructions if you do not have an M1 Mac. We strongly recommend that you use one of the provided Virtual Machines (VM)...">
<!-- JS --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://kit.fontawesome.com/6ecbd6c532.js" crossorigin="anonymous"></script><script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="libs/bootstrap-4.6.0/bootstrap.min.css" rel="stylesheet">
<script src="libs/bootstrap-4.6.0/bootstrap.bundle.min.js"></script><script src="libs/bs3compat-0.5.1.9000/transition.js"></script><script src="libs/bs3compat-0.5.1.9000/tabs.js"></script><script src="libs/bs3compat-0.5.1.9000/bs3compat.js"></script><link href="libs/bs4_book-1.0.0/bs4_book.css" rel="stylesheet">
<script src="libs/bs4_book-1.0.0/bs4_book.js"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- CSS --><style type="text/css">
    
    div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  </style>
<link rel="stylesheet" href="style.css">
</head>
<body data-spy="scroll" data-target="#toc">

<div class="container-fluid">
<div class="row">
  <header class="col-sm-12 col-lg-3 sidebar sidebar-book"><a class="sr-only sr-only-focusable" href="#content">Skip to main content</a>

    <div class="d-flex align-items-start justify-content-between">
      <h1>
        <a href="index.html" title="">scRNAseq Analysis in R with Seurat</a>
      </h1>
      <button class="btn btn-outline-primary d-lg-none ml-2 mt-1" type="button" data-toggle="collapse" data-target="#main-nav" aria-expanded="true" aria-controls="main-nav"><i class="fas fa-bars"></i><span class="sr-only">Show table of contents</span></button>
    </div>

    <div id="main-nav" class="collapse-lg">
      <form role="search">
        <input id="search" class="form-control" type="search" placeholder="Search" aria-label="Search">
</form>

      <nav aria-label="Table of contents"><h2>Table of contents</h2>
        <ul class="book-toc list-unstyled">
<li><a class="" href="index.html"><span class="header-section-number">1</span> Getting started</a></li>
<li><a class="" href="set-up.html"><span class="header-section-number">2</span> Set up</a></li>
<li><a class="active" href="m1-installation.html"><span class="header-section-number">3</span> M1 Installation</a></li>
<li><a class="" href="schedule.html"><span class="header-section-number">4</span> Schedule</a></li>
<li class="book-part">Seurat Object</li>
<li><a class="" href="structure.html"><span class="header-section-number">5</span> Structure</a></li>
<li><a class="" href="load.html"><span class="header-section-number">6</span> Load data</a></li>
<li class="book-part">Single Cell Analysis</li>
<li><a class="" href="qc.html"><span class="header-section-number">7</span> QC Filtering</a></li>
<li><a class="" href="norm.html"><span class="header-section-number">8</span> Normalisation</a></li>
<li><a class="" href="reducedims.html"><span class="header-section-number">9</span> PCAs and UMAPs</a></li>
<li><a class="" href="dimensionality-reduction.html"><span class="header-section-number">10</span> Dimensionality reduction</a></li>
<li><a class="" href="clustering.html"><span class="header-section-number">11</span> Clustering</a></li>
<li><a class="" href="clustermarkers.html"><span class="header-section-number">12</span> Cluster Markers</a></li>
<li class="book-part">Futher Analysis</li>
<li><a class="" href="singler.html"><span class="header-section-number">13</span> SingleR</a></li>
<li><a class="" href="de2.html"><span class="header-section-number">14</span> Differential Expression</a></li>
<li><a class="" href="CellCycle.html"><span class="header-section-number">15</span> Cell cycle Assignment</a></li>
<li><a class="" href="Harmony.html"><span class="header-section-number">16</span> Data set integration with Harmony</a></li>
<li class="book-part">Other Resources</li>
<li><a class="" href="resources.html"><span class="header-section-number">17</span> Resources</a></li>
<li><a class="" href="solutions.html"><span class="header-section-number">18</span> Solutions</a></li>
<li><a class="" href="acknowledgements.html"><span class="header-section-number">19</span> Acknowledgements</a></li>
<li><a class="" href="session-info.html"><span class="header-section-number">20</span> Session info</a></li>
</ul>

        <div class="book-extra">
          <p><a id="book-repo" href="https://github.com/rstudio/bookdown-demo">View book source <i class="fab fa-github"></i></a></p>
        </div>
      </nav>
</div>
  </header><main class="col-sm-12 col-md-9 col-lg-7" id="content"><div id="m1-installation" class="section level1" number="3">
<h1>
<span class="header-section-number">3</span> M1 Installation<a class="anchor" aria-label="anchor" href="#m1-installation"><i class="fas fa-link"></i></a>
</h1>
<p>NOTE: <strong>THESE INSTRUCTIONS ARE ONLY FOR M1 MACS, please use the <a href="set-up.html">installation instructions</a> if you do not have an M1 Mac. We strongly recommend that you use one of the provided Virtual Machines (VM) and follow these instructions in your own time. Details on how to log on and access the VM will be provided during the workshop.</strong></p>
<p>How to check if you have an M1 Mac:</p>
<ol style="list-style-type: decimal">
<li><p>Click the <code>About this Mac</code> in the drop down menu from the top left corner of the screen
<img src="images/Select-About-This-Mac-from-Apple-Menu.png"></p></li>
<li><p>If the popup displays <code>Chip Apple M1</code>(see image below) - you have an M1 Mac computer with all its assorted problems with R package installations and you might need the instructions below to troubleshoot installing the packages required for this workshop. However, we strongly recommend using a VM for the workshop and to try these instructions in your own time</p></li>
</ol>
<div class="inline-figure"><img src="images/Chip-Section-in-About-This-Mac.png"></div>
<p>Alternatively if you see something like this:</p>
<div class="inline-figure"><img src="images/About-This-Mac-Intel.png"></div>
<p>You shouldn’t have too much grief getting the packages installed and should follow these <a href="set-up.html">installation instructions.</a></p>
<div id="instructions-that-should-work" class="section level2" number="3.1">
<h2>
<span class="header-section-number">3.1</span> Instructions That Should Work<a class="anchor" aria-label="anchor" href="#instructions-that-should-work"><i class="fas fa-link"></i></a>
</h2>
<p>From the installation instructions, the steps below work on R.4.2</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html">install.packages</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"Seurat"</span>, <span class="st">"dplyr"</span>, <span class="st">"remotes"</span>, <span class="st">"R.utils"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html">install.packages</a></span><span class="op">(</span><span class="st">"BiocManager"</span><span class="op">)</span></span>
<span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://bioconductor.github.io/BiocManager/reference/install.html">install</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">'SingleR'</span>, <span class="st">'celldex'</span>,</span>
<span>                       <span class="st">'BiocGenerics'</span>, <span class="st">'DelayedArray'</span>, <span class="st">'DelayedMatrixStats'</span>,</span>
<span>                       <span class="st">'limma'</span>, <span class="st">'S4Vectors'</span>, <span class="st">'SingleCellExperiment'</span>,</span>
<span>                       <span class="st">'SummarizedExperiment'</span>, <span class="st">'batchelor'</span>, <span class="st">'Matrix.utils'</span><span class="op">)</span>,</span>
<span>                     update<span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">remotes</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html">install_github</a></span><span class="op">(</span><span class="st">"plotly/plotly.R"</span><span class="op">)</span></span>
<span><span class="fu">remotes</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html">install_github</a></span><span class="op">(</span><span class="st">"pfh/langevitour"</span><span class="op">)</span></span></code></pre></div>
</div>
<div id="instructions-that-will-fail" class="section level2" number="3.2">
<h2>
<span class="header-section-number">3.2</span> Instructions That Will Fail<a class="anchor" aria-label="anchor" href="#instructions-that-will-fail"><i class="fas fa-link"></i></a>
</h2>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">remotes</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_version.html">install_version</a></span><span class="op">(</span><span class="st">"RcppArmadillo"</span>, version<span class="op">=</span><span class="st">"0.11.2.4.0"</span><span class="op">)</span></span>
<span><span class="fu">remotes</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_version.html">install_version</a></span><span class="op">(</span><span class="st">"harmony"</span>, version<span class="op">=</span><span class="st">"0.1"</span>, upgrade<span class="op">=</span><span class="st">"never"</span><span class="op">)</span></span></code></pre></div>
</div>
<div id="troubleshooting-installation-on-m1-mac-with-r.4.2" class="section level2" number="3.3">
<h2>
<span class="header-section-number">3.3</span> Troubleshooting installation on M1 Mac with R.4.2:<a class="anchor" aria-label="anchor" href="#troubleshooting-installation-on-m1-mac-with-r.4.2"><i class="fas fa-link"></i></a>
</h2>
<ol style="list-style-type: decimal">
<li><p>Do not install any packages from source - select <code>no</code> when asked. Try to install and see if the packages will loads</p></li>
<li><p>Seurat installation fails and needs additional packages to install to load Seurat:</p></li>
</ol>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html">install.packages</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"rgeos"</span>, <span class="st">"spatstat.core"</span>, <span class="st">"RcppEigen"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<ol start="3" style="list-style-type: decimal">
<li>Harmony will not install and returns an error that looks like something like this:</li>
</ol>
<pre><code>ld: warning: directory not found for option '-L/opt/R/arm64/gfortran/lib/gcc/aarch64-apple-darwin20.6.0/12.0.1'
ld: library not found for -lemutls_w
clang: error: linker command failed with exit code 1 (use -v to see invocation)
make: *** [harmony.so] Error 1
ERROR: compilation failed for package ‘harmony’
* removing ‘/Library/Frameworks/R.framework/Versions/4.2-arm64/Resources/library/harmony’
Warning message:
In i.p(...) :
  installation of package ‘/var/folders/1k/bbv_y78s4zj0m29mb6fwj3km69dgdb/T//Rtmp9jOOMQ/file83ee3a886f7c/harmony_0.1.1.tar.gz’ had non-zero exit status</code></pre>
<p>OR:</p>
<pre><code>1 warning and 1 error generated.
make: *** [harmony.o] Error 1
ERROR: compilation failed for package ‘harmony’
* removing ‘/Library/Frameworks/R.framework/Versions/4.2-arm64/Resources/library/harmony’
Warning message:
In i.p(...) :
  installation of package ‘/var/folders/1k/bbv_y78s4zj0m29mb6fwj3km69dgdb/T//Rtmpl9BbIi/remotes15d0d7fc6b490/harmony’ had non-zero exit status</code></pre>
<p>Solution: install <a href="https://mac.r-project.org/tools/">GNU Fortran compiler</a>. Following the instructions for that link in regards to Apple Silicon Macs (M1 and above):</p>
<ol style="list-style-type: lower-roman">
<li>Download the experimental branch of GNU Fortran (run the code below in your terminal)</li>
</ol>
<div class="sourceCode" id="cb8"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb8-1"><a href="m1-installation.html#cb8-1" tabindex="-1"></a><span class="fu">wget</span> https://mac.r-project.org/tools/gfortran-12.0.1-20220312-is-darwin20-arm64.tar.xz</span></code></pre></div>
<ol start="2" style="list-style-type: lower-roman">
<li>This binary unpacks into /opt/R/arm64/gfortran plus a symbolic link in /opt/R/arm64/bin/gfortran. To unpack use:</li>
</ol>
<div class="sourceCode" id="cb9"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb9-1"><a href="m1-installation.html#cb9-1" tabindex="-1"></a><span class="fu">tar</span> fxz gfortran-12.0.1-20220312-is-darwin20-arm64.tar.xz <span class="at">-C</span> /</span></code></pre></div>
<p>If this fails because you do not have permission, run it with sudo as so:</p>
<div class="sourceCode" id="cb10"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb10-1"><a href="m1-installation.html#cb10-1" tabindex="-1"></a><span class="fu">sudo</span> tar fxz gfortran-12.0.1-20220312-is-darwin20-arm64.tar.xz <span class="at">-C</span> /</span></code></pre></div>
<p>It will ask for a password - this should be the same password as your user account.</p>
<ol start="3" style="list-style-type: lower-roman">
<li>Now attempt to install harmony:</li>
</ol>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">remotes</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_version.html">install_version</a></span><span class="op">(</span><span class="st">"RcppArmadillo"</span>, version<span class="op">=</span><span class="st">"0.11.2.4.0"</span><span class="op">)</span></span>
<span><span class="fu">remotes</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_version.html">install_version</a></span><span class="op">(</span><span class="st">"harmony"</span>, version<span class="op">=</span><span class="st">"0.1"</span>, upgrade<span class="op">=</span><span class="st">"never"</span><span class="op">)</span></span></code></pre></div>

</div>
</div>
  <div class="chapter-nav">
<div class="prev"><a href="set-up.html"><span class="header-section-number">2</span> Set up</a></div>
<div class="next"><a href="schedule.html"><span class="header-section-number">4</span> Schedule</a></div>
</div></main><div class="col-md-3 col-lg-2 d-none d-md-block sidebar sidebar-chapter">
    <nav id="toc" data-toggle="toc" aria-label="On this page"><h2>On this page</h2>
      <ul class="nav navbar-nav">
<li><a class="nav-link" href="#m1-installation"><span class="header-section-number">3</span> M1 Installation</a></li>
<li><a class="nav-link" href="#instructions-that-should-work"><span class="header-section-number">3.1</span> Instructions That Should Work</a></li>
<li><a class="nav-link" href="#instructions-that-will-fail"><span class="header-section-number">3.2</span> Instructions That Will Fail</a></li>
<li><a class="nav-link" href="#troubleshooting-installation-on-m1-mac-with-r.4.2"><span class="header-section-number">3.3</span> Troubleshooting installation on M1 Mac with R.4.2:</a></li>
</ul>

      <div class="book-extra">
        <ul class="list-unstyled">
<li><a id="book-source" href="https://github.com/rstudio/bookdown-demo/blob/master/01-02-macos_installation.Rmd">View source <i class="fab fa-github"></i></a></li>
          <li><a id="book-edit" href="https://github.com/rstudio/bookdown-demo/edit/master/01-02-macos_installation.Rmd">Edit this page <i class="fab fa-github"></i></a></li>
        </ul>
</div>
    </nav>
</div>

</div>
</div> <!-- .container -->

<footer class="bg-primary text-light mt-5"><div class="container"><div class="row">

  <div class="col-12 col-md-6 mt-3">
    <p>"<strong>scRNAseq Analysis in R with Seurat</strong>" was written by . It was last built on Compiled: November 14, 2023.</p>
  </div>

  <div class="col-12 col-md-6 mt-3">
    <p>This book was built by the <a class="text-light" href="https://bookdown.org">bookdown</a> R package.</p>
  </div>

</div></div>
</footer>
</body>
</html>
