<!DOCTYPE html>
<html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Chapter 19 Session info | scRNAseq Analysis in R with Seurat</title>
<meta name="author" content="Monash Genomics and Bioinformatics Platform (MGBP)">
<meta name="description" content="library(pander) demo.Rmd_session &lt;- sessionInfo() pander(demo.Rmd_session) R version 4.3.2 (2023-10-31) Platform: aarch64-apple-darwin20 (64-bit) locale:...">
<meta name="generator" content="bookdown 0.36 with bs4_book()">
<meta property="og:title" content="Chapter 19 Session info | scRNAseq Analysis in R with Seurat">
<meta property="og:type" content="book">
<meta property="og:description" content="library(pander) demo.Rmd_session &lt;- sessionInfo() pander(demo.Rmd_session) R version 4.3.2 (2023-10-31) Platform: aarch64-apple-darwin20 (64-bit) locale:...">
<meta name="twitter:card" content="summary">
<meta name="twitter:title" content="Chapter 19 Session info | scRNAseq Analysis in R with Seurat">
<meta name="twitter:description" content="library(pander) demo.Rmd_session &lt;- sessionInfo() pander(demo.Rmd_session) R version 4.3.2 (2023-10-31) Platform: aarch64-apple-darwin20 (64-bit) locale:...">
<!-- JS --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://kit.fontawesome.com/6ecbd6c532.js" crossorigin="anonymous"></script><script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="libs/bootstrap-4.6.0/bootstrap.min.css" rel="stylesheet">
<script src="libs/bootstrap-4.6.0/bootstrap.bundle.min.js"></script><script src="libs/bs3compat-0.5.1.9000/transition.js"></script><script src="libs/bs3compat-0.5.1.9000/tabs.js"></script><script src="libs/bs3compat-0.5.1.9000/bs3compat.js"></script><link href="libs/bs4_book-1.0.0/bs4_book.css" rel="stylesheet">
<script src="libs/bs4_book-1.0.0/bs4_book.js"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- CSS --><style type="text/css">
    
    div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  </style>
<link rel="stylesheet" href="style.css">
</head>
<body data-spy="scroll" data-target="#toc">

<div class="container-fluid">
<div class="row">
  <header class="col-sm-12 col-lg-3 sidebar sidebar-book"><a class="sr-only sr-only-focusable" href="#content">Skip to main content</a>

    <div class="d-flex align-items-start justify-content-between">
      <h1>
        <a href="index.html" title="">scRNAseq Analysis in R with Seurat</a>
      </h1>
      <button class="btn btn-outline-primary d-lg-none ml-2 mt-1" type="button" data-toggle="collapse" data-target="#main-nav" aria-expanded="true" aria-controls="main-nav"><i class="fas fa-bars"></i><span class="sr-only">Show table of contents</span></button>
    </div>

    <div id="main-nav" class="collapse-lg">
      <form role="search">
        <input id="search" class="form-control" type="search" placeholder="Search" aria-label="Search">
</form>

      <nav aria-label="Table of contents"><h2>Table of contents</h2>
        <ul class="book-toc list-unstyled">
<li><a class="" href="index.html"><span class="header-section-number">1</span> Getting started</a></li>
<li><a class="" href="set-up.html"><span class="header-section-number">2</span> Set up</a></li>
<li><a class="" href="schedule.html"><span class="header-section-number">3</span> Schedule</a></li>
<li class="book-part">Seurat Object</li>
<li><a class="" href="structure.html"><span class="header-section-number">4</span> Structure</a></li>
<li><a class="" href="load.html"><span class="header-section-number">5</span> Load data</a></li>
<li class="book-part">Single Cell Analysis</li>
<li><a class="" href="qc.html"><span class="header-section-number">6</span> QC Filtering</a></li>
<li><a class="" href="norm.html"><span class="header-section-number">7</span> Normalisation</a></li>
<li><a class="" href="reducedims.html"><span class="header-section-number">8</span> PCAs and UMAPs</a></li>
<li><a class="" href="dimensionality-reduction.html"><span class="header-section-number">9</span> Dimensionality reduction</a></li>
<li><a class="" href="clustering.html"><span class="header-section-number">10</span> Clustering</a></li>
<li><a class="" href="clustermarkers.html"><span class="header-section-number">11</span> Cluster Markers</a></li>
<li class="book-part">Futher Analysis</li>
<li><a class="" href="singler.html"><span class="header-section-number">12</span> SingleR</a></li>
<li><a class="" href="de2.html"><span class="header-section-number">13</span> Differential Expression</a></li>
<li><a class="" href="CellCycle.html"><span class="header-section-number">14</span> Cell cycle Assignment</a></li>
<li><a class="" href="Harmony.html"><span class="header-section-number">15</span> Data set integration with Harmony</a></li>
<li class="book-part">Other Resources</li>
<li><a class="" href="resources.html"><span class="header-section-number">16</span> Resources</a></li>
<li><a class="" href="solutions.html"><span class="header-section-number">17</span> Solutions</a></li>
<li><a class="" href="acknowledgements.html"><span class="header-section-number">18</span> Acknowledgements</a></li>
<li><a class="active" href="session-info.html"><span class="header-section-number">19</span> Session info</a></li>
</ul>

        <div class="book-extra">
          <p><a id="book-repo" href="https://github.com/MonashBioinformaticsPlatform/Nov2023_scRNAseq_workshop">View book source <i class="fab fa-github"></i></a></p>
        </div>
      </nav>
</div>
  </header><main class="col-sm-12 col-md-9 col-lg-7" id="content"><div id="session-info" class="section level1" number="19">
<h1>
<span class="header-section-number">19</span> Session info<a class="anchor" aria-label="anchor" href="#session-info"><i class="fas fa-link"></i></a>
</h1>
<div class="sourceCode" id="cb109"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://rapporter.github.io/pander/">pander</a></span><span class="op">)</span></span>
<span><span class="va">demo.Rmd_session</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/pander/man/pander.html">pander</a></span><span class="op">(</span><span class="va">demo.Rmd_session</span><span class="op">)</span></span></code></pre></div>
<p><strong>R version 4.3.2 (2023-10-31)</strong></p>
<p><strong>Platform:</strong> aarch64-apple-darwin20 (64-bit)</p>
<p><strong>locale:</strong>
en_US.UTF-8||en_US.UTF-8||en_US.UTF-8||C||en_US.UTF-8||en_US.UTF-8</p>
<p><strong>attached base packages:</strong>
<em>stats4</em>, <em>stats</em>, <em>graphics</em>, <em>grDevices</em>, <em>utils</em>, <em>datasets</em>, <em>methods</em> and <em>base</em></p>
<p><strong>other attached packages:</strong>
<em>pander(v.0.6.5)</em>, <em>harmony(v.1.1.0)</em>, <em>Rcpp(v.1.0.11)</em>, <em>edgeR(v.4.0.1)</em>, <em>limma(v.3.58.1)</em>, <em>celldex(v.1.12.0)</em>, <em>SingleR(v.2.4.0)</em>, <em>SingleCellExperiment(v.1.24.0)</em>, <em>SummarizedExperiment(v.1.32.0)</em>, <em>Biobase(v.2.62.0)</em>, <em>GenomicRanges(v.1.54.1)</em>, <em>GenomeInfoDb(v.1.38.1)</em>, <em>IRanges(v.2.36.0)</em>, <em>S4Vectors(v.0.40.1)</em>, <em>BiocGenerics(v.0.48.1)</em>, <em>MatrixGenerics(v.1.14.0)</em>, <em>matrixStats(v.1.1.0)</em>, <em>clustree(v.0.5.0)</em>, <em>ggraph(v.2.1.0)</em>, <em>patchwork(v.1.1.3)</em>, <em>ggplot2(v.3.4.4)</em>, <em>dplyr(v.1.1.3)</em>, <em>SeuratObject(v.5.0.0)</em> and <em>Seurat(v.4.4.0)</em></p>
<p><strong>loaded via a namespace (and not attached):</strong>
<em>fs(v.1.6.3)</em>, <em>spatstat.sparse(v.3.0-3)</em>, <em>bitops(v.1.0-7)</em>, <em>httr(v.1.4.7)</em>, <em>RColorBrewer(v.1.1-3)</em>, <em>tools(v.4.3.2)</em>, <em>sctransform(v.0.4.1)</em>, <em>backports(v.1.4.1)</em>, <em>utf8(v.1.2.4)</em>, <em>R6(v.2.5.1)</em>, <em>lazyeval(v.0.2.2)</em>, <em>uwot(v.0.1.16)</em>, <em>withr(v.2.5.2)</em>, <em>sp(v.2.1-1)</em>, <em>gridExtra(v.2.3)</em>, <em>downlit(v.0.4.3)</em>, <em>progressr(v.0.14.0)</em>, <em>cli(v.3.6.1)</em>, <em>textshaping(v.0.3.7)</em>, <em>spatstat.explore(v.3.2-5)</em>, <em>labeling(v.0.4.3)</em>, <em>sass(v.0.4.7)</em>, <em>spatstat.data(v.3.0-3)</em>, <em>ggridges(v.0.5.4)</em>, <em>pbapply(v.1.7-2)</em>, <em>systemfonts(v.1.0.5)</em>, <em>R.utils(v.2.12.2)</em>, <em>parallelly(v.1.36.0)</em>, <em>rstudioapi(v.0.15.0)</em>, <em>RSQLite(v.2.3.3)</em>, <em>generics(v.0.1.3)</em>, <em>ica(v.1.0-3)</em>, <em>spatstat.random(v.3.2-1)</em>, <em>Matrix(v.1.6-1.1)</em>, <em>ggbeeswarm(v.0.7.2)</em>, <em>fansi(v.1.0.5)</em>, <em>abind(v.1.4-5)</em>, <em>R.methodsS3(v.1.8.2)</em>, <em>lifecycle(v.1.0.4)</em>, <em>yaml(v.2.3.7)</em>, <em>SparseArray(v.1.2.2)</em>, <em>BiocFileCache(v.2.10.1)</em>, <em>Rtsne(v.0.16)</em>, <em>grid(v.4.3.2)</em>, <em>blob(v.1.2.4)</em>, <em>promises(v.1.2.1)</em>, <em>ExperimentHub(v.2.10.0)</em>, <em>crayon(v.1.5.2)</em>, <em>miniUI(v.0.1.1.1)</em>, <em>lattice(v.0.21-9)</em>, <em>beachmat(v.2.18.0)</em>, <em>cowplot(v.1.1.1)</em>, <em>KEGGREST(v.1.42.0)</em>, <em>pillar(v.1.9.0)</em>, <em>knitr(v.1.45)</em>, <em>future.apply(v.1.11.0)</em>, <em>codetools(v.0.2-19)</em>, <em>leiden(v.0.4.3)</em>, <em>glue(v.1.6.2)</em>, <em>data.table(v.1.14.8)</em>, <em>vctrs(v.0.6.4)</em>, <em>png(v.0.1-8)</em>, <em>spam(v.2.10-0)</em>, <em>gtable(v.0.3.4)</em>, <em>cachem(v.1.0.8)</em>, <em>xfun(v.0.41)</em>, <em>S4Arrays(v.1.2.0)</em>, <em>mime(v.0.12)</em>, <em>tidygraph(v.1.2.3)</em>, <em>survival(v.3.5-7)</em>, <em>statmod(v.1.5.0)</em>, <em>interactiveDisplayBase(v.1.40.0)</em>, <em>ellipsis(v.0.3.2)</em>, <em>fitdistrplus(v.1.1-11)</em>, <em>ROCR(v.1.0-11)</em>, <em>nlme(v.3.1-163)</em>, <em>bit64(v.4.0.5)</em>, <em>filelock(v.1.0.2)</em>, <em>RcppAnnoy(v.0.0.21)</em>, <em>bslib(v.0.5.1.9000)</em>, <em>irlba(v.2.3.5.1)</em>, <em>vipor(v.0.4.5)</em>, <em>KernSmooth(v.2.23-22)</em>, <em>colorspace(v.2.1-0)</em>, <em>DBI(v.1.1.3)</em>, <em>ggrastr(v.1.0.2)</em>, <em>tidyselect(v.1.2.0)</em>, <em>bit(v.4.0.5)</em>, <em>compiler(v.4.3.2)</em>, <em>curl(v.5.1.0)</em>, <em>xml2(v.1.3.5)</em>, <em>DelayedArray(v.0.28.0)</em>, <em>plotly(v.4.10.3.9000)</em>, <em>bookdown(v.0.36)</em>, <em>checkmate(v.2.3.0)</em>, <em>scales(v.1.2.1)</em>, <em>lmtest(v.0.9-40)</em>, <em>rappdirs(v.0.3.3)</em>, <em>stringr(v.1.5.0)</em>, <em>digest(v.0.6.33)</em>, <em>goftest(v.1.2-3)</em>, <em>spatstat.utils(v.3.0-4)</em>, <em>rmarkdown(v.2.25)</em>, <em>RhpcBLASctl(v.0.23-42)</em>, <em>XVector(v.0.42.0)</em>, <em>htmltools(v.0.5.7.9000)</em>, <em>pkgconfig(v.2.0.3)</em>, <em>sparseMatrixStats(v.1.14.0)</em>, <em>highr(v.0.10)</em>, <em>dbplyr(v.2.4.0)</em>, <em>fastmap(v.1.1.1)</em>, <em>rlang(v.1.1.2)</em>, <em>htmlwidgets(v.1.6.2)</em>, <em>shiny(v.1.7.5.1)</em>, <em>DelayedMatrixStats(v.1.24.0)</em>, <em>farver(v.2.1.1)</em>, <em>jquerylib(v.0.1.4)</em>, <em>zoo(v.1.8-12)</em>, <em>jsonlite(v.1.8.7)</em>, <em>BiocParallel(v.1.36.0)</em>, <em>R.oo(v.1.25.0)</em>, <em>BiocSingular(v.1.18.0)</em>, <em>RCurl(v.1.98-1.13)</em>, <em>magrittr(v.2.0.3)</em>, <em>GenomeInfoDbData(v.1.2.11)</em>, <em>dotCall64(v.1.1-0)</em>, <em>munsell(v.0.5.0)</em>, <em>viridis(v.0.6.4)</em>, <em>reticulate(v.1.34.0)</em>, <em>stringi(v.1.7.12)</em>, <em>zlibbioc(v.1.48.0)</em>, <em>MASS(v.7.3-60)</em>, <em>AnnotationHub(v.3.10.0)</em>, <em>plyr(v.1.8.9)</em>, <em>parallel(v.4.3.2)</em>, <em>listenv(v.0.9.0)</em>, <em>ggrepel(v.0.9.4)</em>, <em>deldir(v.1.0-9)</em>, <em>Biostrings(v.2.70.1)</em>, <em>graphlayouts(v.1.0.2)</em>, <em>splines(v.4.3.2)</em>, <em>tensor(v.1.5)</em>, <em>locfit(v.1.5-9.8)</em>, <em>igraph(v.1.5.1)</em>, <em>spatstat.geom(v.3.2-7)</em>, <em>reshape2(v.1.4.4)</em>, <em>ScaledMatrix(v.1.10.0)</em>, <em>BiocVersion(v.3.18.0)</em>, <em>evaluate(v.0.23)</em>, <em>BiocManager(v.1.30.22)</em>, <em>tweenr(v.2.0.2)</em>, <em>httpuv(v.1.6.12)</em>, <em>RANN(v.2.6.1)</em>, <em>tidyr(v.1.3.0)</em>, <em>purrr(v.1.0.2)</em>, <em>polyclip(v.1.10-6)</em>, <em>future(v.1.33.0)</em>, <em>scattermore(v.1.2)</em>, <em>ggforce(v.0.4.1)</em>, <em>rsvd(v.1.0.5)</em>, <em>xtable(v.1.8-4)</em>, <em>later(v.1.3.1)</em>, <em>viridisLite(v.0.4.2)</em>, <em>ragg(v.1.2.6)</em>, <em>tibble(v.3.2.1)</em>, <em>memoise(v.2.0.1)</em>, <em>beeswarm(v.0.4.0)</em>, <em>AnnotationDbi(v.1.64.1)</em>, <em>cluster(v.2.1.4)</em> and <em>globals(v.0.16.2)</em></p>

</div>
  <div class="chapter-nav">
<div class="prev"><a href="acknowledgements.html"><span class="header-section-number">18</span> Acknowledgements</a></div>
<div class="empty"></div>
</div></main><div class="col-md-3 col-lg-2 d-none d-md-block sidebar sidebar-chapter">
    <nav id="toc" data-toggle="toc" aria-label="On this page"><h2>On this page</h2>
      <ul class="nav navbar-nav"><li><a class="nav-link" href="#session-info"><span class="header-section-number">19</span> Session info</a></li></ul>

      <div class="book-extra">
        <ul class="list-unstyled">
<li><a id="book-source" href="https://github.com/MonashBioinformaticsPlatform/Nov2023_scRNAseq_workshop/blob/master/08-session_info.Rmd">View source <i class="fab fa-github"></i></a></li>
          <li><a id="book-edit" href="https://github.com/MonashBioinformaticsPlatform/Nov2023_scRNAseq_workshop/edit/master/08-session_info.Rmd">Edit this page <i class="fab fa-github"></i></a></li>
        </ul>
</div>
    </nav>
</div>

</div>
</div> <!-- .container -->

<footer class="bg-primary text-light mt-5"><div class="container"><div class="row">

  <div class="col-12 col-md-6 mt-3">
    <p>"<strong>scRNAseq Analysis in R with Seurat</strong>" was written by Monash Genomics and Bioinformatics Platform (MGBP). It was last built on Compiled: November 16, 2023.</p>
  </div>

  <div class="col-12 col-md-6 mt-3">
    <p>This book was built by the <a class="text-light" href="https://bookdown.org">bookdown</a> R package.</p>
  </div>

</div></div>
</footer>
</body>
</html>
