<!DOCTYPE html>
<html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Chapter 2 Set up | scRNAseq Analysis in R with Seurat</title>
<meta name="author" content="Monash Genomics and Bioinformatics Platform (MGBP)">
<meta name="description" content="For this workshop you will have available a VM. However, if you which to use your local computer here are the set up instructions: IMPORTANT: If you have an M1 Mac - Make sure you have gfortan. ...">
<meta name="generator" content="bookdown 0.36 with bs4_book()">
<meta property="og:title" content="Chapter 2 Set up | scRNAseq Analysis in R with Seurat">
<meta property="og:type" content="book">
<meta property="og:description" content="For this workshop you will have available a VM. However, if you which to use your local computer here are the set up instructions: IMPORTANT: If you have an M1 Mac - Make sure you have gfortan. ...">
<meta name="twitter:card" content="summary">
<meta name="twitter:title" content="Chapter 2 Set up | scRNAseq Analysis in R with Seurat">
<meta name="twitter:description" content="For this workshop you will have available a VM. However, if you which to use your local computer here are the set up instructions: IMPORTANT: If you have an M1 Mac - Make sure you have gfortan. ...">
<!-- JS --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://kit.fontawesome.com/6ecbd6c532.js" crossorigin="anonymous"></script><script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="libs/bootstrap-4.6.0/bootstrap.min.css" rel="stylesheet">
<script src="libs/bootstrap-4.6.0/bootstrap.bundle.min.js"></script><script src="libs/bs3compat-0.5.1.9000/transition.js"></script><script src="libs/bs3compat-0.5.1.9000/tabs.js"></script><script src="libs/bs3compat-0.5.1.9000/bs3compat.js"></script><link href="libs/bs4_book-1.0.0/bs4_book.css" rel="stylesheet">
<script src="libs/bs4_book-1.0.0/bs4_book.js"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- CSS --><style type="text/css">
    
    div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  </style>
<link rel="stylesheet" href="style.css">
</head>
<body data-spy="scroll" data-target="#toc">

<div class="container-fluid">
<div class="row">
  <header class="col-sm-12 col-lg-3 sidebar sidebar-book"><a class="sr-only sr-only-focusable" href="#content">Skip to main content</a>

    <div class="d-flex align-items-start justify-content-between">
      <h1>
        <a href="index.html" title="">scRNAseq Analysis in R with Seurat</a>
      </h1>
      <button class="btn btn-outline-primary d-lg-none ml-2 mt-1" type="button" data-toggle="collapse" data-target="#main-nav" aria-expanded="true" aria-controls="main-nav"><i class="fas fa-bars"></i><span class="sr-only">Show table of contents</span></button>
    </div>

    <div id="main-nav" class="collapse-lg">
      <form role="search">
        <input id="search" class="form-control" type="search" placeholder="Search" aria-label="Search">
</form>

      <nav aria-label="Table of contents"><h2>Table of contents</h2>
        <ul class="book-toc list-unstyled">
<li><a class="" href="index.html"><span class="header-section-number">1</span> Getting started</a></li>
<li><a class="active" href="set-up.html"><span class="header-section-number">2</span> Set up</a></li>
<li><a class="" href="schedule.html"><span class="header-section-number">3</span> Schedule</a></li>
<li class="book-part">Seurat Object</li>
<li><a class="" href="structure.html"><span class="header-section-number">4</span> Structure</a></li>
<li><a class="" href="load.html"><span class="header-section-number">5</span> Load data</a></li>
<li class="book-part">Single Cell Analysis</li>
<li><a class="" href="qc.html"><span class="header-section-number">6</span> QC Filtering</a></li>
<li><a class="" href="norm.html"><span class="header-section-number">7</span> Normalisation</a></li>
<li><a class="" href="reducedims.html"><span class="header-section-number">8</span> PCAs and UMAPs</a></li>
<li><a class="" href="dimensionality-reduction.html"><span class="header-section-number">9</span> Dimensionality reduction</a></li>
<li><a class="" href="clustering.html"><span class="header-section-number">10</span> Clustering</a></li>
<li><a class="" href="clustermarkers.html"><span class="header-section-number">11</span> Cluster Markers</a></li>
<li class="book-part">Futher Analysis</li>
<li><a class="" href="singler.html"><span class="header-section-number">12</span> SingleR</a></li>
<li><a class="" href="de2.html"><span class="header-section-number">13</span> Differential Expression</a></li>
<li><a class="" href="CellCycle.html"><span class="header-section-number">14</span> Cell cycle Assignment</a></li>
<li><a class="" href="Harmony.html"><span class="header-section-number">15</span> Data set integration with Harmony</a></li>
<li class="book-part">Other Resources</li>
<li><a class="" href="resources.html"><span class="header-section-number">16</span> Resources</a></li>
<li><a class="" href="solutions.html"><span class="header-section-number">17</span> Solutions</a></li>
<li><a class="" href="acknowledgements.html"><span class="header-section-number">18</span> Acknowledgements</a></li>
<li><a class="" href="session-info.html"><span class="header-section-number">19</span> Session info</a></li>
</ul>

        <div class="book-extra">
          <p><a id="book-repo" href="https://github.com/MonashBioinformaticsPlatform/Nov2023_scRNAseq_workshop">View book source <i class="fab fa-github"></i></a></p>
        </div>
      </nav>
</div>
  </header><main class="col-sm-12 col-md-9 col-lg-7" id="content"><div id="set-up" class="section level1" number="2">
<h1>
<span class="header-section-number">2</span> Set up<a class="anchor" aria-label="anchor" href="#set-up"><i class="fas fa-link"></i></a>
</h1>
<p>For this workshop you will have available a VM. However, if you which to use your local computer here are the set up instructions:</p>
<p><strong>IMPORTANT: If you have an M1 Mac - Make sure you have <a href="https://mac.r-project.org/tools/">gfortan</a>.</strong></p>
<div id="get-the-workshop-material-and-data" class="section level2" number="2.1">
<h2>
<span class="header-section-number">2.1</span> Get the workshop material and data<a class="anchor" aria-label="anchor" href="#get-the-workshop-material-and-data"><i class="fas fa-link"></i></a>
</h2>
<p>In RStudio <strong>create a new project</strong>. This ensures all the files for this workshop are placed in their own folder.</p>
<p>Once youâ€™ve created a new project, run the following R code</p>
</div>
<div id="package-installation" class="section level2" number="2.2">
<h2>
<span class="header-section-number">2.2</span> Package Installation<a class="anchor" aria-label="anchor" href="#package-installation"><i class="fas fa-link"></i></a>
</h2>
<p>For this workshop, several packages need to be installed.</p>
<p>BiocManager likes to update installed packages, but we have disabled this in the R code below. If your installation fails, then you might need to turn updates on. Note that if there are a large number of packages that BiocManager wants to update it can take several hours.</p>
<p>These instructions have been tested with R version 4.2.1 and Bioconductor version 3.16.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Install required packages for Seurat:</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html">install.packages</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"Seurat"</span>, <span class="st">"dplyr"</span>, <span class="st">"remotes"</span>, <span class="st">"R.utils"</span>,<span class="st">"harmony"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Install required Bioconductor packages</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html">install.packages</a></span><span class="op">(</span><span class="st">"BiocManager"</span><span class="op">)</span></span>
<span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://bioconductor.github.io/BiocManager/reference/install.html">install</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">'SingleR'</span>, <span class="st">'celldex'</span>,</span>
<span>                       <span class="st">'BiocGenerics'</span>, <span class="st">'DelayedArray'</span>, <span class="st">'DelayedMatrixStats'</span>,</span>
<span>                       <span class="st">'limma'</span>, <span class="st">'S4Vectors'</span>, <span class="st">'SingleCellExperiment'</span>,</span>
<span>                       <span class="st">'SummarizedExperiment'</span>,<span class="st">'edgeR'</span> <span class="op">)</span>,</span>
<span>                     update<span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span></span>
<span></span>
<span></span>
<span><span class="co">## install clustree</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html">install.packages</a></span><span class="op">(</span><span class="st">"clustree"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html">install.packages</a></span><span class="op">(</span><span class="st">"RColorBrewer"</span><span class="op">)</span></span></code></pre></div>
</div>
<div id="raw-data" class="section level2" number="2.3">
<h2>
<span class="header-section-number">2.3</span> Raw Data<a class="anchor" aria-label="anchor" href="#raw-data"><i class="fas fa-link"></i></a>
</h2>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># where are you? what folder are you working in</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/getwd.html">getwd</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "/Users/lper0012/Dropbox/BioPlatform/Teaching/Nov2023_scRNAseq_workshop"</span></span>
<span></span>
<span><span class="co">## Download and untar the data</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/options.html">options</a></span><span class="op">(</span>timeout<span class="op">=</span><span class="fl">3600</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/download.file.html">download.file</a></span><span class="op">(</span></span>
<span>    <span class="st">"https://bioinformatics.erc.monash.edu/home/lper0012/SingleCellWorkshopData/data.tar"</span>,<span class="st">"data.tar"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/untar.html">untar</a></span><span class="op">(</span><span class="st">"data.tar"</span><span class="op">)</span></span></code></pre></div>

</div>
</div>
  <div class="chapter-nav">
<div class="prev"><a href="index.html"><span class="header-section-number">1</span> Getting started</a></div>
<div class="next"><a href="schedule.html"><span class="header-section-number">3</span> Schedule</a></div>
</div></main><div class="col-md-3 col-lg-2 d-none d-md-block sidebar sidebar-chapter">
    <nav id="toc" data-toggle="toc" aria-label="On this page"><h2>On this page</h2>
      <ul class="nav navbar-nav">
<li><a class="nav-link" href="#set-up"><span class="header-section-number">2</span> Set up</a></li>
<li><a class="nav-link" href="#get-the-workshop-material-and-data"><span class="header-section-number">2.1</span> Get the workshop material and data</a></li>
<li><a class="nav-link" href="#package-installation"><span class="header-section-number">2.2</span> Package Installation</a></li>
<li><a class="nav-link" href="#raw-data"><span class="header-section-number">2.3</span> Raw Data</a></li>
</ul>

      <div class="book-extra">
        <ul class="list-unstyled">
<li><a id="book-source" href="https://github.com/MonashBioinformaticsPlatform/Nov2023_scRNAseq_workshop/blob/master/01-01-setup.Rmd">View source <i class="fab fa-github"></i></a></li>
          <li><a id="book-edit" href="https://github.com/MonashBioinformaticsPlatform/Nov2023_scRNAseq_workshop/edit/master/01-01-setup.Rmd">Edit this page <i class="fab fa-github"></i></a></li>
        </ul>
</div>
    </nav>
</div>

</div>
</div> <!-- .container -->

<footer class="bg-primary text-light mt-5"><div class="container"><div class="row">

  <div class="col-12 col-md-6 mt-3">
    <p>"<strong>scRNAseq Analysis in R with Seurat</strong>" was written by Monash Genomics and Bioinformatics Platform (MGBP). It was last built on Compiled: November 16, 2023.</p>
  </div>

  <div class="col-12 col-md-6 mt-3">
    <p>This book was built by the <a class="text-light" href="https://bookdown.org">bookdown</a> R package.</p>
  </div>

</div></div>
</footer>
</body>
</html>
